// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/UploadFilePickerTab/UploadFilePickerTab.module.scss.js
/* tslint:disable */
require("./UploadFilePickerTab.module.css");
var styles = {
    filePicker: 'filePicker_88d26ec5',
    tabsContainer: 'tabsContainer_88d26ec5',
    tabContainer: 'tabContainer_88d26ec5',
    nav: 'nav_88d26ec5',
    tabHeaderContainer: 'tabHeaderContainer_88d26ec5',
    breadcrumbNav: 'breadcrumbNav_88d26ec5',
    breadcrumbNavItem: 'breadcrumbNavItem_88d26ec5',
    tab: 'tab_88d26ec5',
    tabOffset: 'tabOffset_88d26ec5',
    tabFiles: 'tabFiles_88d26ec5',
    scrollablePaneWrapper: 'scrollablePaneWrapper_88d26ec5',
    actionButtonsContainer: 'actionButtonsContainer_88d26ec5',
    actionButtons: 'actionButtons_88d26ec5',
    actionButton: 'actionButton_88d26ec5',
    tabHeader: 'tabHeader_88d26ec5',
    selectedTab: 'selectedTab_88d26ec5',
    header: 'header_88d26ec5',
    xlg: 'xlg_88d26ec5',
    xxlg: 'xxlg_88d26ec5',
    xxxlg: 'xxxlg_88d26ec5',
    xxxxlg: 'xxxxlg_88d26ec5',
    itemPickerTopBar: 'itemPickerTopBar_88d26ec5',
    commandBarNoChevron: 'commandBarNoChevron_88d26ec5',
    fileItem: 'fileItem_88d26ec5',
    folderItem: 'folderItem_88d26ec5',
    fileTypeIcon: 'fileTypeIcon_88d26ec5',
    fileTypeIconIcon: 'fileTypeIconIcon_88d26ec5',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_88d26ec5',
    iconColumnHeader: 'iconColumnHeader_88d26ec5',
    fileRow: 'fileRow_88d26ec5',
    folderRow: 'folderRow_88d26ec5',
    emptyFolder: 'emptyFolder_88d26ec5',
    emptyFolderImage: 'emptyFolderImage_88d26ec5',
    emptyFolderImageTag: 'emptyFolderImageTag_88d26ec5',
    emptyFolderTitle: 'emptyFolderTitle_88d26ec5',
    emptyFolderSubText: 'emptyFolderSubText_88d26ec5',
    emptyFolderMobile: 'emptyFolderMobile_88d26ec5',
    emptyFolderPc: 'emptyFolderPc_88d26ec5',
    emptyFolderIcon: 'emptyFolderIcon_88d26ec5',
    localTabSinglePreview: 'localTabSinglePreview_88d26ec5',
    localTabSinglePreviewImage: 'localTabSinglePreviewImage_88d26ec5',
    localTabFilename: 'localTabFilename_88d26ec5',
    localTabLabel: 'localTabLabel_88d26ec5',
    localTabInput: 'localTabInput_88d26ec5'
};
export default styles;
/* tslint:enable */
