// @microsoft/sp-page-context/lib-commonjs/index.js
"use strict";
/**
 * Page context services for the SharePoint Framework
 *
 * @remarks
 * The page context allows an application and its web parts to share common
 * data about the current page, such as its contents, URL, list, user, permissions,
 * navigation menu, etc.
 *
 * @packagedocumentation
 */
Object.defineProperty(exports, "__esModule", { value: true });
var Page_1 = require("./Page");
Object.defineProperty(exports, "Page", { enumerable: true, get: function () { return Page_1.default; } });
var PageContext_1 = require("./PageContext");
Object.defineProperty(exports, "PageContext", { enumerable: true, get: function () { return PageContext_1.default; } });
var SPWeb_1 = require("./SPWeb");
Object.defineProperty(exports, "SPWeb", { enumerable: true, get: function () { return SPWeb_1.default; } });
var SPSite_1 = require("./SPSite");
Object.defineProperty(exports, "SPSite", { enumerable: true, get: function () { return SPSite_1.default; } });
var SPUser_1 = require("./SPUser");
Object.defineProperty(exports, "SPUser", { enumerable: true, get: function () { return SPUser_1.default; } });
var SPField_1 = require("./SPField");
Object.defineProperty(exports, "SPField", { enumerable: true, get: function () { return SPField_1.default; } });
var SPList_1 = require("./SPList");
Object.defineProperty(exports, "SPList", { enumerable: true, get: function () { return SPList_1.default; } });
var SPListItem_1 = require("./SPListItem");
Object.defineProperty(exports, "SPListItem", { enumerable: true, get: function () { return SPListItem_1.default; } });
var SPFeatureInfo_1 = require("./SPFeatureInfo");
Object.defineProperty(exports, "_SPFeatureInfo", { enumerable: true, get: function () { return SPFeatureInfo_1.default; } });
var SPDayOfWeek_1 = require("./SPDayOfWeek");
Object.defineProperty(exports, "SPDayOfWeek", { enumerable: true, get: function () { return SPDayOfWeek_1.SPDayOfWeek; } });
var CultureInfo_1 = require("./CultureInfo");
Object.defineProperty(exports, "CultureInfo", { enumerable: true, get: function () { return CultureInfo_1.default; } });
var O365GroupAssociation_1 = require("./O365GroupAssociation");
Object.defineProperty(exports, "O365GroupAssociation", { enumerable: true, get: function () { return O365GroupAssociation_1.default; } });
/* IO365GroupAssociation,*/
Object.defineProperty(exports, "O365GroupAssociationType", { enumerable: true, get: function () { return O365GroupAssociation_1.O365GroupAssociationType; } });
var SPPermission_1 = require("./SPPermission");
Object.defineProperty(exports, "SPPermission", { enumerable: true, get: function () { return SPPermission_1.default; } });
var SharePointPageContextDataProvider_1 = require("./pageManager/SharePointPageContextDataProvider");
Object.defineProperty(exports, "SharePointPageContextDataProvider", { enumerable: true, get: function () { return SharePointPageContextDataProvider_1.default; } });
var SPTimeZone_1 = require("./SPTimeZone");
Object.defineProperty(exports, "SPTimeZone", { enumerable: true, get: function () { return SPTimeZone_1.default; } });
var AzureActiveDirectoryInfo_1 = require("./AzureActiveDirectoryInfo");
Object.defineProperty(exports, "AzureActiveDirectoryInfo", { enumerable: true, get: function () { return AzureActiveDirectoryInfo_1.default; } });
