// @microsoft/rush-lib/lib/logic/CredentialCache.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CredentialCache = void 0;
const path = __importStar(require("path"));
const node_core_library_1 = require("@rushstack/node-core-library");
const Utilities_1 = require("../utilities/Utilities");
const RushUserConfiguration_1 = require("../api/RushUserConfiguration");
const CACHE_FILENAME = 'credentials.json';
const LATEST_CREDENTIALS_JSON_VERSION = '0.1.0';
/**
 * @beta
 */
class CredentialCache {
    constructor(cacheFilePath, loadedJson, lockfile) {
        this._modified = false;
        this._disposed = false;
        if (loadedJson && loadedJson.version !== LATEST_CREDENTIALS_JSON_VERSION) {
            throw new Error(`Unexpected credentials.json file version: ${loadedJson.version}`);
        }
        this._cacheFilePath = cacheFilePath;
        this._cacheEntries = new Map(Object.entries((loadedJson === null || loadedJson === void 0 ? void 0 : loadedJson.cacheEntries) || {}));
        this._supportsEditing = !!lockfile;
        this._lockfile = lockfile;
    }
    static async initializeAsync(options) {
        const rushUserFolderPath = RushUserConfiguration_1.RushUserConfiguration.getRushUserFolderPath();
        const cacheFilePath = path.join(rushUserFolderPath, CACHE_FILENAME);
        const jsonSchema = node_core_library_1.JsonSchema.fromFile(path.resolve(__dirname, '..', 'schemas', 'credentials.schema.json'));
        let loadedJson;
        try {
            loadedJson = await node_core_library_1.JsonFile.loadAndValidateAsync(cacheFilePath, jsonSchema);
        }
        catch (e) {
            if (!node_core_library_1.FileSystem.isErrnoException(e)) {
                throw e;
            }
        }
        let lockfile;
        if (options.supportEditing) {
            lockfile = await node_core_library_1.LockFile.acquire(rushUserFolderPath, `${CACHE_FILENAME}.lock`);
        }
        const credentialCache = new CredentialCache(cacheFilePath, loadedJson, lockfile);
        return credentialCache;
    }
    static async usingAsync(options, doActionAsync) {
        await Utilities_1.Utilities.usingAsync(async () => await CredentialCache.initializeAsync(options), doActionAsync);
    }
    setCacheEntry(cacheId, credential, expires) {
        this._validate(true);
        const expiresMilliseconds = (expires === null || expires === void 0 ? void 0 : expires.getTime()) || 0;
        const existingCacheEntry = this._cacheEntries.get(cacheId);
        if ((existingCacheEntry === null || existingCacheEntry === void 0 ? void 0 : existingCacheEntry.credential) !== credential ||
            (existingCacheEntry === null || existingCacheEntry === void 0 ? void 0 : existingCacheEntry.expires) !== expiresMilliseconds) {
            this._modified = true;
            this._cacheEntries.set(cacheId, {
                expires: expiresMilliseconds,
                credential
            });
        }
    }
    tryGetCacheEntry(cacheId) {
        this._validate(false);
        const cacheEntry = this._cacheEntries.get(cacheId);
        if (cacheEntry) {
            const result = {
                expires: cacheEntry.expires ? new Date(cacheEntry.expires) : undefined,
                credential: cacheEntry.credential
            };
            return result;
        }
        else {
            return undefined;
        }
    }
    deleteCacheEntry(cacheId) {
        this._validate(true);
        if (this._cacheEntries.has(cacheId)) {
            this._modified = true;
            this._cacheEntries.delete(cacheId);
        }
    }
    trimExpiredEntries() {
        this._validate(true);
        const now = Date.now();
        for (const [cacheId, cacheEntry] of this._cacheEntries.entries()) {
            if (cacheEntry.expires < now) {
                this._cacheEntries.delete(cacheId);
                this._modified = true;
            }
        }
    }
    async saveIfModifiedAsync() {
        this._validate(true);
        if (this._modified) {
            const cacheEntriesJson = {};
            for (const [cacheId, cacheEntry] of this._cacheEntries.entries()) {
                cacheEntriesJson[cacheId] = cacheEntry;
            }
            const newJson = {
                version: LATEST_CREDENTIALS_JSON_VERSION,
                cacheEntries: cacheEntriesJson
            };
            await node_core_library_1.JsonFile.saveAsync(newJson, this._cacheFilePath, {
                ensureFolderExists: true,
                updateExistingFile: true
            });
            this._modified = false;
        }
    }
    dispose() {
        var _a;
        (_a = this._lockfile) === null || _a === void 0 ? void 0 : _a.release();
        this._disposed = true;
    }
    _validate(requiresEditing) {
        if (!this._supportsEditing && requiresEditing) {
            throw new Error(`This instance of ${CredentialCache.name} does not support editing.`);
        }
        if (this._disposed) {
            throw new Error(`This instance of ${CredentialCache.name} has been disposed.`);
        }
    }
}
exports.CredentialCache = CredentialCache;
