// @microsoft/rush-lib/lib/logic/NodeJsCompatibility.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeJsCompatibility = void 0;
const colors_1 = __importDefault(require("colors"));
const semver = __importStar(require("semver"));
/**
 * This constant is the major version of the next LTS node Node.js release. This constant should be updated when
 * a new LTS version is added to Rush's support matrix.
 *
 * LTS schedule: https://nodejs.org/en/about/releases/
 * LTS versions: https://nodejs.org/en/download/releases/
 */
const UPCOMING_NODE_LTS_VERSION = 16;
const nodeVersion = process.versions.node;
const nodeMajorVersion = semver.major(nodeVersion);
/**
 * This class provides useful functions for warning if the current Node.js runtime isn't supported.
 *
 * @internal
 */
class NodeJsCompatibility {
    static warnAboutCompatibilityIssues(options) {
        // Only show the first warning
        return (NodeJsCompatibility.warnAboutVersionTooOld() ||
            NodeJsCompatibility.warnAboutVersionTooNew(options) ||
            NodeJsCompatibility.warnAboutOddNumberedVersion() ||
            NodeJsCompatibility.warnAboutNonLtsVersion(options.rushConfiguration));
    }
    static warnAboutVersionTooOld() {
        if (semver.satisfies(nodeVersion, '< 10.13.0')) {
            // We are on an ancient version of Node.js that is known not to work with Rush
            console.error(colors_1.default.red(`Your version of Node.js (${nodeVersion}) is very old and incompatible with Rush. ` +
                `Please upgrade to the latest Long-Term Support (LTS) version.`));
            return true;
        }
        else {
            return false;
        }
    }
    static warnAboutVersionTooNew(options) {
        if (!options.alreadyReportedNodeTooNewError && nodeMajorVersion >= UPCOMING_NODE_LTS_VERSION + 1) {
            // We are on a much newer release than we have tested and support
            if (options.isRushLib) {
                console.warn(colors_1.default.yellow(`Your version of Node.js (${nodeVersion}) has not been tested with this release ` +
                    `of the Rush engine. Please consider upgrading the "rushVersion" setting in rush.json, ` +
                    `or downgrading Node.js.`));
            }
            else {
                console.warn(colors_1.default.yellow(`Your version of Node.js (${nodeVersion}) has not been tested with this release ` +
                    `of Rush. Please consider installing a newer version of the "@microsoft/rush" ` +
                    `package, or downgrading Node.js.`));
            }
            return true;
        }
        else {
            return false;
        }
    }
    static warnAboutNonLtsVersion(rushConfiguration) {
        if (rushConfiguration && !rushConfiguration.suppressNodeLtsWarning && !NodeJsCompatibility.isLtsVersion) {
            console.warn(colors_1.default.yellow(`Your version of Node.js (${nodeVersion}) is not a Long-Term Support (LTS) release. ` +
                'These versions frequently have bugs. Please consider installing a stable release.'));
            return true;
        }
        else {
            return false;
        }
    }
    static warnAboutOddNumberedVersion() {
        if (NodeJsCompatibility.isOddNumberedVersion) {
            console.warn(colors_1.default.yellow(`Your version of Node.js (${nodeVersion}) is an odd-numbered release. ` +
                `These releases frequently have bugs. Please consider installing a Long Term Support (LTS) ` +
                `version instead.`));
            return true;
        }
        else {
            return false;
        }
    }
    static get isLtsVersion() {
        return !!process.release.lts;
    }
    static get isOddNumberedVersion() {
        return nodeMajorVersion % 2 !== 0;
    }
}
exports.NodeJsCompatibility = NodeJsCompatibility;
