// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/WebSearchTab/WebSearchTab.module.scss.js
/* tslint:disable */
require("./WebSearchTab.module.css");
var styles = {
    filePicker: 'filePicker_4e4adce5',
    tabsContainer: 'tabsContainer_4e4adce5',
    tabContainer: 'tabContainer_4e4adce5',
    nav: 'nav_4e4adce5',
    tabHeaderContainer: 'tabHeaderContainer_4e4adce5',
    breadcrumbNav: 'breadcrumbNav_4e4adce5',
    breadcrumbNavItem: 'breadcrumbNavItem_4e4adce5',
    tab: 'tab_4e4adce5',
    tabOffset: 'tabOffset_4e4adce5',
    tabFiles: 'tabFiles_4e4adce5',
    scrollablePaneWrapper: 'scrollablePaneWrapper_4e4adce5',
    actionButtonsContainer: 'actionButtonsContainer_4e4adce5',
    actionButtons: 'actionButtons_4e4adce5',
    actionButton: 'actionButton_4e4adce5',
    tabHeader: 'tabHeader_4e4adce5',
    selectedTab: 'selectedTab_4e4adce5',
    header: 'header_4e4adce5',
    xlg: 'xlg_4e4adce5',
    xxlg: 'xxlg_4e4adce5',
    xxxlg: 'xxxlg_4e4adce5',
    xxxxlg: 'xxxxlg_4e4adce5',
    itemPickerTopBar: 'itemPickerTopBar_4e4adce5',
    commandBarNoChevron: 'commandBarNoChevron_4e4adce5',
    fileItem: 'fileItem_4e4adce5',
    folderItem: 'folderItem_4e4adce5',
    fileTypeIcon: 'fileTypeIcon_4e4adce5',
    fileTypeIconIcon: 'fileTypeIconIcon_4e4adce5',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_4e4adce5',
    iconColumnHeader: 'iconColumnHeader_4e4adce5',
    fileRow: 'fileRow_4e4adce5',
    folderRow: 'folderRow_4e4adce5',
    emptyFolder: 'emptyFolder_4e4adce5',
    emptyFolderImage: 'emptyFolderImage_4e4adce5',
    emptyFolderImageTag: 'emptyFolderImageTag_4e4adce5',
    emptyFolderTitle: 'emptyFolderTitle_4e4adce5',
    emptyFolderSubText: 'emptyFolderSubText_4e4adce5',
    emptyFolderMobile: 'emptyFolderMobile_4e4adce5',
    emptyFolderPc: 'emptyFolderPc_4e4adce5',
    emptyFolderIcon: 'emptyFolderIcon_4e4adce5',
    searchBoxContainer: 'searchBoxContainer_4e4adce5',
    searchBoxMedium: 'searchBoxMedium_4e4adce5',
    searchBox: 'searchBox_4e4adce5',
    dropdownContainer: 'dropdownContainer_4e4adce5',
    filterDropdown: 'filterDropdown_4e4adce5',
    filePickerFolderCardGrid: 'filePickerFolderCardGrid_4e4adce5',
    filePickerFolderCardTile: 'filePickerFolderCardTile_4e4adce5',
    filePickerFolderCardPadder: 'filePickerFolderCardPadder_4e4adce5',
    filePickerFolderCardSizer: 'filePickerFolderCardSizer_4e4adce5',
    filePickerFolderCardImage: 'filePickerFolderCardImage_4e4adce5',
    filePickerFolderCardLabel: 'filePickerFolderCardLabel_4e4adce5',
    bingGridListCell: 'bingGridListCell_4e4adce5',
    bingTile: 'bingTile_4e4adce5',
    bingTileContent: 'bingTileContent_4e4adce5',
    bingTileFrame: 'bingTileFrame_4e4adce5',
    bingTileCheckCircle: 'bingTileCheckCircle_4e4adce5',
    bingTileNamePlate: 'bingTileNamePlate_4e4adce5',
    isSelected: 'isSelected_4e4adce5',
    copyrightLabel: 'copyrightLabel_4e4adce5',
    noResultLabel: 'noResultLabel_4e4adce5',
    bingTileThumbnail: 'bingTileThumbnail_4e4adce5',
    bingGrildList: 'bingGrildList_4e4adce5'
};
export default styles;
/* tslint:enable */
