// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/controls/FolderTile/FolderTile.module.scss.js
/* tslint:disable */
require("./FolderTile.module.css");
var styles = {
    filePicker: 'filePicker_295c09da',
    tabsContainer: 'tabsContainer_295c09da',
    tabContainer: 'tabContainer_295c09da',
    nav: 'nav_295c09da',
    tabHeaderContainer: 'tabHeaderContainer_295c09da',
    breadcrumbNav: 'breadcrumbNav_295c09da',
    breadcrumbNavItem: 'breadcrumbNavItem_295c09da',
    tab: 'tab_295c09da',
    tabOffset: 'tabOffset_295c09da',
    tabFiles: 'tabFiles_295c09da',
    scrollablePaneWrapper: 'scrollablePaneWrapper_295c09da',
    actionButtonsContainer: 'actionButtonsContainer_295c09da',
    actionButtons: 'actionButtons_295c09da',
    actionButton: 'actionButton_295c09da',
    tabHeader: 'tabHeader_295c09da',
    selectedTab: 'selectedTab_295c09da',
    header: 'header_295c09da',
    xlg: 'xlg_295c09da',
    xxlg: 'xxlg_295c09da',
    xxxlg: 'xxxlg_295c09da',
    xxxxlg: 'xxxxlg_295c09da',
    itemPickerTopBar: 'itemPickerTopBar_295c09da',
    commandBarNoChevron: 'commandBarNoChevron_295c09da',
    fileItem: 'fileItem_295c09da',
    folderItem: 'folderItem_295c09da',
    fileTypeIcon: 'fileTypeIcon_295c09da',
    fileTypeIconIcon: 'fileTypeIconIcon_295c09da',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_295c09da',
    iconColumnHeader: 'iconColumnHeader_295c09da',
    fileRow: 'fileRow_295c09da',
    folderRow: 'folderRow_295c09da',
    emptyFolder: 'emptyFolder_295c09da',
    emptyFolderImage: 'emptyFolderImage_295c09da',
    emptyFolderImageTag: 'emptyFolderImageTag_295c09da',
    emptyFolderTitle: 'emptyFolderTitle_295c09da',
    emptyFolderSubText: 'emptyFolderSubText_295c09da',
    emptyFolderMobile: 'emptyFolderMobile_295c09da',
    emptyFolderPc: 'emptyFolderPc_295c09da',
    emptyFolderIcon: 'emptyFolderIcon_295c09da',
    aboveNameplate: 'aboveNameplate_295c09da',
    tileContent: 'tileContent_295c09da',
    foreground: 'foreground_295c09da',
    grid: 'grid_295c09da',
    cellContent: 'cellContent_295c09da',
    content: 'content_295c09da',
    tile: 'tile_295c09da',
    invokable: 'invokable_295c09da',
    namePlate: 'namePlate_295c09da',
    nameplate: 'nameplate_295c09da',
    name: 'name_295c09da',
    isLarge: 'isLarge_295c09da',
    activity: 'activity_295c09da',
    signalField: 'signalField_295c09da',
    compact: 'compact_295c09da',
    signalFieldValue: 'signalFieldValue_295c09da',
    folderList: 'folderList_295c09da',
    link: 'link_295c09da',
    cell: 'cell_295c09da',
    listCell: 'listCell_295c09da',
    label: 'label_295c09da',
    check: 'check_295c09da',
    selectable: 'selectable_295c09da',
    showCheck: 'showCheck_295c09da',
    selected: 'selected_295c09da',
    dialogMainOverride: 'dialogMainOverride_295c09da',
    itemTileBlankCover: 'itemTileBlankCover_295c09da',
    metadata: 'metadata_295c09da',
    isSmall: 'isSmall_295c09da',
    listPage: 'listPage_295c09da',
    odItemTile2FolderCover: 'odItemTile2FolderCover_295c09da',
    standaloneListBreadcrumb: 'standaloneListBreadcrumb_295c09da',
    folderCover: 'folderCover_295c09da',
    isSmall_88d29bd4: 'isSmall_88d29bd4_295c09da',
    folderCoverBack: 'folderCoverBack_295c09da',
    folderCoverFront: 'folderCoverFront_295c09da',
    folderCoverFrame: 'folderCoverFrame_295c09da',
    folderCoverContent: 'folderCoverContent_295c09da'
};
export default styles;
/* tslint:enable */
