// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/StockImagesTab/StockImages.module.scss.js
/* tslint:disable */
require("./StockImages.module.css");
var styles = {
    filePicker: 'filePicker_f3a3adb6',
    tabsContainer: 'tabsContainer_f3a3adb6',
    tabContainer: 'tabContainer_f3a3adb6',
    nav: 'nav_f3a3adb6',
    tabHeaderContainer: 'tabHeaderContainer_f3a3adb6',
    breadcrumbNav: 'breadcrumbNav_f3a3adb6',
    breadcrumbNavItem: 'breadcrumbNavItem_f3a3adb6',
    tab: 'tab_f3a3adb6',
    tabOffset: 'tabOffset_f3a3adb6',
    tabFiles: 'tabFiles_f3a3adb6',
    scrollablePaneWrapper: 'scrollablePaneWrapper_f3a3adb6',
    actionButtonsContainer: 'actionButtonsContainer_f3a3adb6',
    actionButtons: 'actionButtons_f3a3adb6',
    actionButton: 'actionButton_f3a3adb6',
    tabHeader: 'tabHeader_f3a3adb6',
    selectedTab: 'selectedTab_f3a3adb6',
    header: 'header_f3a3adb6',
    xlg: 'xlg_f3a3adb6',
    xxlg: 'xxlg_f3a3adb6',
    xxxlg: 'xxxlg_f3a3adb6',
    xxxxlg: 'xxxxlg_f3a3adb6',
    itemPickerTopBar: 'itemPickerTopBar_f3a3adb6',
    commandBarNoChevron: 'commandBarNoChevron_f3a3adb6',
    fileItem: 'fileItem_f3a3adb6',
    folderItem: 'folderItem_f3a3adb6',
    fileTypeIcon: 'fileTypeIcon_f3a3adb6',
    fileTypeIconIcon: 'fileTypeIconIcon_f3a3adb6',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_f3a3adb6',
    iconColumnHeader: 'iconColumnHeader_f3a3adb6',
    fileRow: 'fileRow_f3a3adb6',
    folderRow: 'folderRow_f3a3adb6',
    emptyFolder: 'emptyFolder_f3a3adb6',
    emptyFolderImage: 'emptyFolderImage_f3a3adb6',
    emptyFolderImageTag: 'emptyFolderImageTag_f3a3adb6',
    emptyFolderTitle: 'emptyFolderTitle_f3a3adb6',
    emptyFolderSubText: 'emptyFolderSubText_f3a3adb6',
    emptyFolderMobile: 'emptyFolderMobile_f3a3adb6',
    emptyFolderPc: 'emptyFolderPc_f3a3adb6',
    emptyFolderIcon: 'emptyFolderIcon_f3a3adb6',
    stockImagesPickerContainer: 'stockImagesPickerContainer_f3a3adb6',
    stockImagesPicker: 'stockImagesPicker_f3a3adb6'
};
export default styles;
/* tslint:enable */
