// @microsoft/sp-webpart-base/lib/core/IIframedWebPartMessage.js
/**
 * Strings for the 'action' property on IIframedWebPartMessage.
 * @internal
 */
// tslint:disable-next-line:typedef variable-name
export var IframedWebPartAction = {
    /**
     * An action to delete the web part.
     */
    DeleteWebPart: 'iframeDeleteWebPart',
    /**
     * An action to request the current DisplayMode of a WebPart.
     */
    RequestDisplayMode: 'iframeRequestDisplayMode',
    /**
     * An action requested on the PropertyPane.
     */
    RequestPropertyPaneAction: 'iframeRequestPropertyPaneAction',
    /**
     * An action to set the DisplayMode of an iframed WebPart.
     */
    SetDisplayMode: 'iframeSetDisplayMode',
    /**
     * An action to set the dimensions of the iframe to match it's embed contents.
     * Example dimensions: height and/or width.
     */
    SetDimensions: 'iframeSetDimensions',
    /**
     * An action from the IframedWebPartController to the window that
     * is showing the WebPart or PropertyPane.
     */
    SetWebPartData: 'iframeSetWebPartData',
    /**
     * An action to update the WebPart with the data changed by the PropertyPane.
     */
    UpdateWebPartData: 'iframeUpdateWebPartData',
    /**
     * An action to update the PropertyPane with the data changed by the rendered WebPart.
     */
    UpdatePropertyPaneData: 'iframeUpdatePropertyPaneData',
    /**
     * The action corresponds to a PropertyPane lifecycle event.
     */
    UpdatePropertyPaneLifeCycle: 'iframePropertyPaneLifeCycle',
    /**
     * Action to indicate that the web part is rendered in the iframe application
     * where the property pane is to be rendered.
     */
    WebPartRenderedInPropertyPaneIframe: 'webPartRenderedInPropertyPaneIframe',
    /**
     * Action to communicate a message to the main page from the dynamic data scenarios.
     */
    DynamicDataMessage: 'isolatedDynamicDataMessage',
    /**
     * Action corresponds to update life cycle method in page host
     */
    UpdateWebpartLifeCycle: 'iframeUpdateWebpartLifeCycle'
};
