// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/LinkFilePickerTab/LinkFilePickerTab.module.scss.js
/* tslint:disable */
require("./LinkFilePickerTab.module.css");
var styles = {
    filePicker: 'filePicker_8d5d089d',
    tabsContainer: 'tabsContainer_8d5d089d',
    tabContainer: 'tabContainer_8d5d089d',
    nav: 'nav_8d5d089d',
    tabHeaderContainer: 'tabHeaderContainer_8d5d089d',
    breadcrumbNav: 'breadcrumbNav_8d5d089d',
    breadcrumbNavItem: 'breadcrumbNavItem_8d5d089d',
    tab: 'tab_8d5d089d',
    tabOffset: 'tabOffset_8d5d089d',
    tabFiles: 'tabFiles_8d5d089d',
    scrollablePaneWrapper: 'scrollablePaneWrapper_8d5d089d',
    actionButtonsContainer: 'actionButtonsContainer_8d5d089d',
    actionButtons: 'actionButtons_8d5d089d',
    actionButton: 'actionButton_8d5d089d',
    tabHeader: 'tabHeader_8d5d089d',
    selectedTab: 'selectedTab_8d5d089d',
    header: 'header_8d5d089d',
    xlg: 'xlg_8d5d089d',
    xxlg: 'xxlg_8d5d089d',
    xxxlg: 'xxxlg_8d5d089d',
    xxxxlg: 'xxxxlg_8d5d089d',
    itemPickerTopBar: 'itemPickerTopBar_8d5d089d',
    commandBarNoChevron: 'commandBarNoChevron_8d5d089d',
    fileItem: 'fileItem_8d5d089d',
    folderItem: 'folderItem_8d5d089d',
    fileTypeIcon: 'fileTypeIcon_8d5d089d',
    fileTypeIconIcon: 'fileTypeIconIcon_8d5d089d',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_8d5d089d',
    iconColumnHeader: 'iconColumnHeader_8d5d089d',
    fileRow: 'fileRow_8d5d089d',
    folderRow: 'folderRow_8d5d089d',
    emptyFolder: 'emptyFolder_8d5d089d',
    emptyFolderImage: 'emptyFolderImage_8d5d089d',
    emptyFolderImageTag: 'emptyFolderImageTag_8d5d089d',
    emptyFolderTitle: 'emptyFolderTitle_8d5d089d',
    emptyFolderSubText: 'emptyFolderSubText_8d5d089d',
    emptyFolderMobile: 'emptyFolderMobile_8d5d089d',
    emptyFolderPc: 'emptyFolderPc_8d5d089d',
    emptyFolderIcon: 'emptyFolderIcon_8d5d089d',
    linkTextField: 'linkTextField_8d5d089d',
    localTabSinglePreview: 'localTabSinglePreview_8d5d089d',
    localTabSinglePreviewImage: 'localTabSinglePreviewImage_8d5d089d'
};
export default styles;
/* tslint:enable */
