// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/OneDriveFilesTab/OneDriveFilesTab.module.scss.js
/* tslint:disable */
require("./OneDriveFilesTab.module.css");
var styles = {
    filePicker: 'filePicker_61f7bf33',
    tabsContainer: 'tabsContainer_61f7bf33',
    tabContainer: 'tabContainer_61f7bf33',
    nav: 'nav_61f7bf33',
    tabHeaderContainer: 'tabHeaderContainer_61f7bf33',
    breadcrumbNav: 'breadcrumbNav_61f7bf33',
    breadcrumbNavItem: 'breadcrumbNavItem_61f7bf33',
    tab: 'tab_61f7bf33',
    tabOffset: 'tabOffset_61f7bf33',
    tabFiles: 'tabFiles_61f7bf33',
    scrollablePaneWrapper: 'scrollablePaneWrapper_61f7bf33',
    actionButtonsContainer: 'actionButtonsContainer_61f7bf33',
    actionButtons: 'actionButtons_61f7bf33',
    actionButton: 'actionButton_61f7bf33',
    tabHeader: 'tabHeader_61f7bf33',
    selectedTab: 'selectedTab_61f7bf33',
    header: 'header_61f7bf33',
    xlg: 'xlg_61f7bf33',
    xxlg: 'xxlg_61f7bf33',
    xxxlg: 'xxxlg_61f7bf33',
    xxxxlg: 'xxxxlg_61f7bf33',
    itemPickerTopBar: 'itemPickerTopBar_61f7bf33',
    commandBarNoChevron: 'commandBarNoChevron_61f7bf33',
    fileItem: 'fileItem_61f7bf33',
    folderItem: 'folderItem_61f7bf33',
    fileTypeIcon: 'fileTypeIcon_61f7bf33',
    fileTypeIconIcon: 'fileTypeIconIcon_61f7bf33',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_61f7bf33',
    iconColumnHeader: 'iconColumnHeader_61f7bf33',
    fileRow: 'fileRow_61f7bf33',
    folderRow: 'folderRow_61f7bf33',
    emptyFolder: 'emptyFolder_61f7bf33',
    emptyFolderImage: 'emptyFolderImage_61f7bf33',
    emptyFolderImageTag: 'emptyFolderImageTag_61f7bf33',
    emptyFolderTitle: 'emptyFolderTitle_61f7bf33',
    emptyFolderSubText: 'emptyFolderSubText_61f7bf33',
    emptyFolderMobile: 'emptyFolderMobile_61f7bf33',
    emptyFolderPc: 'emptyFolderPc_61f7bf33',
    emptyFolderIcon: 'emptyFolderIcon_61f7bf33'
};
export default styles;
/* tslint:enable */
