// @microsoft/sp-webpart-base/lib-commonjs/index.js
"use strict";
/**
 * SharePoint Framework support for building web parts.
 *
 * @remarks
 * This package defines the APIs used by developers to create a custom web part.
 * A web part is a reusable visual object that a page author can add to their content,
 * and customize using a property pane.  Examples of web parts include an embedded
 * video player, a map, a group calendar, a chart, etc.
 *
 * @packagedocumentation
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var BaseWebPartContext_1 = require("./core/BaseWebPartContext");
Object.defineProperty(exports, "BaseWebPartContext", { enumerable: true, get: function () { return BaseWebPartContext_1.default; } });
var WebPartContext_1 = require("./core/WebPartContext");
Object.defineProperty(exports, "WebPartContext", { enumerable: true, get: function () { return WebPartContext_1.default; } });
var IWebPartData_1 = require("./core/IWebPartData");
Object.defineProperty(exports, "WebPartIsolationLevel", { enumerable: true, get: function () { return IWebPartData_1.WebPartIsolationLevel; } });
var BaseWebPartHost_1 = require("./components/host/BaseWebPartHost");
Object.defineProperty(exports, "_BaseWebPartHost", { enumerable: true, get: function () { return BaseWebPartHost_1.BaseWebPartHost; } });
var MinimalWebPartHost_1 = require("./components/host/MinimalWebPartHost");
Object.defineProperty(exports, "_MinimalWebPartHost", { enumerable: true, get: function () { return MinimalWebPartHost_1.MinimalWebPartHost; } });
var TeamsAwareWebPartHost_1 = require("./components/host/TeamsAwareWebPartHost");
Object.defineProperty(exports, "_TeamsAwareWebPartHost", { enumerable: true, get: function () { return TeamsAwareWebPartHost_1.TeamsAwareWebPartHost; } });
var BaseWebPart_1 = require("./core/BaseWebPart");
Object.defineProperty(exports, "BaseWebPart", { enumerable: true, get: function () { return BaseWebPart_1.default; } });
var BaseClientSideWebPart_1 = require("./core/BaseClientSideWebPart");
Object.defineProperty(exports, "BaseClientSideWebPart", { enumerable: true, get: function () { return BaseClientSideWebPart_1.default; } });
var ClientSideWebPartManager_1 = require("./core/ClientSideWebPartManager");
Object.defineProperty(exports, "ClientSideWebPartManager", { enumerable: true, get: function () { return ClientSideWebPartManager_1.default; } });
var ClientSideWebPartManagerFactory_1 = require("./core/ClientSideWebPartManagerFactory");
Object.defineProperty(exports, "ClientSideWebPartManagerFactory", { enumerable: true, get: function () { return ClientSideWebPartManagerFactory_1.default; } });
var ClientSideWebPartStatusRenderer_1 = require("./core/ClientSideWebPartStatusRenderer");
Object.defineProperty(exports, "_ClientSideWebPartStatusRenderer", { enumerable: true, get: function () { return ClientSideWebPartStatusRenderer_1.default; } });
var WebPartDataConverter_1 = require("./core/WebPartDataConverter");
Object.defineProperty(exports, "WebPartDataConverter", { enumerable: true, get: function () { return WebPartDataConverter_1.default; } });
var WebPartFormFactor_1 = require("./core/WebPartFormFactor");
Object.defineProperty(exports, "WebPartFormFactor", { enumerable: true, get: function () { return WebPartFormFactor_1.WebPartFormFactor; } });
/* IFramedWebPartController typed exports */
var IIframedWebPartMessage_1 = require("./core/IIframedWebPartMessage");
Object.defineProperty(exports, "_IframedWebPartAction", { enumerable: true, get: function () { return IIframedWebPartMessage_1.IframedWebPartAction; } });
tslib_1.__exportStar(require("./SPPropertyPane"), exports);
var MinimalWebPartContainer_1 = require("./components/container/MinimalWebPartContainer");
Object.defineProperty(exports, "_MinimalWebPartContainer", { enumerable: true, get: function () { return MinimalWebPartContainer_1.default; } });
var PropertyPaneLoader_1 = require("./PropertyPaneLoader");
Object.defineProperty(exports, "_PropertyPaneLoader", { enumerable: true, get: function () { return PropertyPaneLoader_1.default; } });
