// @pnp/spfx-property-controls/lib/propertyFields/columnPicker/PropertyFieldColumnPickerHost.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import * as React from 'react';
import { Dropdown } from 'office-ui-fabric-react/lib/Dropdown';
import { Async } from 'office-ui-fabric-react/lib/Utilities';
import { Label } from 'office-ui-fabric-react/lib/Label';
import { SPColumnPickerService } from '../../services/SPColumnPickerService';
import FieldErrorMessage from '../errorMessage/FieldErrorMessage';
import * as telemetry from '../../common/telemetry';
import { setPropertyValue } from '../../helpers/GeneralHelper';
// Empty column value
var EMPTY_COLUMN_KEY = 'NO_COLUMN_SELECTED';
/**
 * Renders the controls for PropertyFieldColumnPicker component
 */
var PropertyFieldColumnPickerHost = /** @class */ (function (_super) {
    __extends(PropertyFieldColumnPickerHost, _super);
    /**
     * Constructor method
     */
    function PropertyFieldColumnPickerHost(props) {
        var _this = _super.call(this, props) || this;
        _this.options = [];
        telemetry.track('PropertyFieldColumnPicker', {
            disabled: props.disabled
        });
        _this.state = {
            results: _this.options,
            errorMessage: ''
        };
        _this.async = new Async(_this);
        _this.validate = _this.validate.bind(_this);
        _this.onChanged = _this.onChanged.bind(_this);
        _this.notifyAfterValidate = _this.notifyAfterValidate.bind(_this);
        _this.delayedValidate = _this.async.debounce(_this.validate, _this.props.deferredValidationTime);
        return _this;
    }
    PropertyFieldColumnPickerHost.prototype.componentDidMount = function () {
        // Start retrieving the list columns
        this.loadColumns();
    };
    PropertyFieldColumnPickerHost.prototype.componentDidUpdate = function (prevProps, _prevState) {
        if (this.props.listId !== prevProps.listId || this.props.webAbsoluteUrl !== prevProps.webAbsoluteUrl) {
            this.loadColumns();
        }
    };
    /**
     * Loads the columns from a SharePoint list
     */
    PropertyFieldColumnPickerHost.prototype.loadColumns = function () {
        var _this = this;
        var _a = this.props, context = _a.context, columnReturnProperty = _a.columnReturnProperty, selectedColumn = _a.selectedColumn, displayHiddenColumns = _a.displayHiddenColumns;
        var columnService = new SPColumnPickerService(this.props, context);
        var columnsToExclude = this.props.columnsToExclude || [];
        this.options = [];
        columnService.getColumns(displayHiddenColumns).then(function (response) {
            // Start mapping the Columns that are selected
            var value = response.value || [];
            value.forEach(function (column) {
                var colPropsToCheck = columnReturnProperty ? column[columnReturnProperty] : column.Id;
                if (selectedColumn === colPropsToCheck) {
                    _this.selectedKey = columnReturnProperty ? column[columnReturnProperty] : column.Id;
                }
                // Make sure that the current column is NOT in the 'columnsToExclude' array
                if (columnsToExclude.indexOf(column.Title) === -1 && columnsToExclude.indexOf(column.Id) === -1) {
                    _this.options.push({
                        key: columnReturnProperty ? column[columnReturnProperty] : column.Id,
                        text: column.Title
                    });
                }
            });
            // Option to unselect the column
            _this.options.unshift({
                key: EMPTY_COLUMN_KEY,
                text: ''
            });
            // Update the current component state
            _this.setState({
                results: _this.options,
                selectedKey: _this.selectedKey
            });
        });
    };
    /**
     * Raises when a column has been selected
     */
    PropertyFieldColumnPickerHost.prototype.onChanged = function (option, _index) {
        var newValue = option.key;
        this.delayedValidate(newValue);
    };
    /**
     * Validates the new custom field value
     */
    PropertyFieldColumnPickerHost.prototype.validate = function (value) {
        var _this = this;
        if (this.props.onGetErrorMessage === null || this.props.onGetErrorMessage === undefined) {
            this.notifyAfterValidate(this.props.selectedColumn, value);
            return;
        }
        if (this.latestValidateValue === value) {
            return;
        }
        this.latestValidateValue = value;
        var errResult = this.props.onGetErrorMessage(value || '');
        if (typeof errResult !== 'undefined') {
            if (typeof errResult === 'string') {
                if (errResult === '') {
                    this.notifyAfterValidate(this.props.selectedColumn, value);
                }
                this.setState({
                    errorMessage: errResult
                });
            }
            else {
                errResult.then(function (errorMessage) {
                    if (!errorMessage) {
                        _this.notifyAfterValidate(_this.props.selectedColumn, value);
                    }
                    _this.setState({
                        errorMessage: errorMessage
                    });
                });
            }
        }
        else {
            this.notifyAfterValidate(this.props.selectedColumn, value);
        }
    };
    /**
     * Notifies the parent Web Part of a property value change
     */
    PropertyFieldColumnPickerHost.prototype.notifyAfterValidate = function (oldValue, newValue) {
        // Check if the user wanted to unselect the column
        var propValue = newValue === EMPTY_COLUMN_KEY ? '' : newValue;
        // Deselect all options
        this.options = this.state.results.map(function (option) {
            if (option.selected) {
                option.selected = false;
            }
            return option;
        });
        // Set the current selected key
        this.selectedKey = newValue;
        // Update the state
        this.setState({
            selectedKey: this.selectedKey,
            results: this.options
        });
        if (this.props.onPropertyChange && propValue !== null) {
            // Store the new property value
            setPropertyValue(this.props.properties, this.props.targetProperty, propValue);
            // Trigger the default onPropertyChange event
            this.props.onPropertyChange(this.props.targetProperty, oldValue, propValue);
            // Trigger the apply button
            if (typeof this.props.onChange !== 'undefined' && this.props.onChange !== null) {
                this.props.onChange(this.props.targetProperty, propValue);
            }
        }
    };
    /**
     * Called when the component will unmount
     */
    PropertyFieldColumnPickerHost.prototype.componentWillUnmount = function () {
        if (typeof this.async !== 'undefined') {
            this.async.dispose();
        }
    };
    /**
     * Renders the SPColumnPicker controls with Office UI Fabric
     */
    PropertyFieldColumnPickerHost.prototype.render = function () {
        // Renders content
        return (React.createElement("div", null,
            this.props.label && React.createElement(Label, null, this.props.label),
            React.createElement(Dropdown, { disabled: this.props.disabled, label: '', onChanged: this.onChanged, options: this.state.results, selectedKey: this.state.selectedKey }),
            React.createElement(FieldErrorMessage, { errorMessage: this.state.errorMessage })));
    };
    return PropertyFieldColumnPickerHost;
}(React.Component));
export default PropertyFieldColumnPickerHost;
