// @microsoft/sp-page-context/lib/O365GroupAssociation.js
import { Guid } from '@microsoft/sp-core-library';
/**
 * {@inheritDoc  O365GroupAssociation."type"}
 * @alpha
 */
export var O365GroupAssociationType;
(function (O365GroupAssociationType) {
    O365GroupAssociationType[O365GroupAssociationType["Unknown"] = 0] = "Unknown";
    O365GroupAssociationType[O365GroupAssociationType["Private"] = 1] = "Private";
    O365GroupAssociationType[O365GroupAssociationType["Public"] = 2] = "Public";
})(O365GroupAssociationType || (O365GroupAssociationType = {}));
/**
 * This class is primarily used with the `PageContext` class. If the site is associated with an O365
 * group, this object provides contextual information about the group.
 *
 * @alpha
 */
var O365GroupAssociation = /** @class */ (function () {
    /**
     * @internal
     */
    function O365GroupAssociation(data) {
        this.id = Guid.parse(data.id);
        switch (data.type) {
            case 'Private':
                this.type = O365GroupAssociationType.Private;
                break;
            case 'Public':
                this.type = O365GroupAssociationType.Public;
                break;
            default:
                this.type = O365GroupAssociationType.Unknown;
                break;
        }
    }
    return O365GroupAssociation;
}());
export default O365GroupAssociation;
