// @microsoft/sp-page-context/lib/SPSite.js
import { Guid } from '@microsoft/sp-core-library';
import O365GroupAssociation from './O365GroupAssociation';
/**
 * This class is primarily used with the `PageContext` class.  It provides contextual
 * information for the SharePoint site collection ("SPSite") that hosts the page.
 *
 * @public
 */
var SPSite = /** @class */ (function () {
    /**
     * @internal
     */
    function SPSite(data) {
        this.absoluteUrl = data.absoluteUrl;
        this.cdnPrefix = data.cdnPrefix || '';
        this.classification = data.classification;
        this.correlationId = Guid.parse(data.correlationId);
        this.group = data.group ? new O365GroupAssociation(data.group) : undefined;
        this.id = Guid.parse(data.id);
        this.isNoScriptEnabled = data.isNoScriptEnabled;
        this.recycleBinItemCount = data.recycleBinItemCount;
        this.serverRelativeUrl = data.serverRelativeUrl;
        this.serverRequestPath = data.serverRequestPath;
        this.sitePagesEnabled = data.sitePagesEnabled;
        this.sitePagesFeatureVersion = data.sitePagesFeatureVersion ? data.sitePagesFeatureVersion : 0;
    }
    return SPSite;
}());
export default SPSite;
