// @microsoft/sp-page-context/lib/SPWeb.js
import { Guid } from '@microsoft/sp-core-library';
import SPPermission from './SPPermission';
import SPTimeZone from './SPTimeZone';
/**
 * This class is primarily used with the `PageContext` class. It provides contextual
 * information for the SharePoint site ("SPWeb") that hosts the page.
 *
 * @public
 */
var SPWeb = /** @class */ (function () {
    /**
     * @internal
     */
    function SPWeb(data) {
        this.absoluteUrl = data.absoluteUrl;
        this.id = Guid.parse(data.id);
        this.isAppWeb = data.isAppWeb;
        this.language = data.language;
        this.languageName = data.languageName || '';
        this.logoUrl = data.logoUrl;
        this.permissions = new SPPermission(data.permissions);
        this.timeZoneInfo = data.timeZoneInfo ? new SPTimeZone(data.timeZoneInfo) : undefined;
        this.firstDayOfWeek = data.firstDayOfWeek;
        this.serverRelativeUrl = data.serverRelativeUrl;
        this.templateName = data.templateName || '';
        this.title = data.title || '';
        this.description = data.description || '';
    }
    return SPWeb;
}());
export default SPWeb;
