// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/SiteFilePickerTab/SiteFilePickerTab.module.scss.js
/* tslint:disable */
require("./SiteFilePickerTab.module.css");
var styles = {
    filePicker: 'filePicker_94e9e3b0',
    tabsContainer: 'tabsContainer_94e9e3b0',
    tabContainer: 'tabContainer_94e9e3b0',
    nav: 'nav_94e9e3b0',
    tabHeaderContainer: 'tabHeaderContainer_94e9e3b0',
    breadcrumbNav: 'breadcrumbNav_94e9e3b0',
    breadcrumbNavItem: 'breadcrumbNavItem_94e9e3b0',
    tab: 'tab_94e9e3b0',
    tabOffset: 'tabOffset_94e9e3b0',
    tabFiles: 'tabFiles_94e9e3b0',
    scrollablePaneWrapper: 'scrollablePaneWrapper_94e9e3b0',
    actionButtonsContainer: 'actionButtonsContainer_94e9e3b0',
    actionButtons: 'actionButtons_94e9e3b0',
    actionButton: 'actionButton_94e9e3b0',
    tabHeader: 'tabHeader_94e9e3b0',
    selectedTab: 'selectedTab_94e9e3b0',
    header: 'header_94e9e3b0',
    xlg: 'xlg_94e9e3b0',
    xxlg: 'xxlg_94e9e3b0',
    xxxlg: 'xxxlg_94e9e3b0',
    xxxxlg: 'xxxxlg_94e9e3b0',
    itemPickerTopBar: 'itemPickerTopBar_94e9e3b0',
    commandBarNoChevron: 'commandBarNoChevron_94e9e3b0',
    fileItem: 'fileItem_94e9e3b0',
    folderItem: 'folderItem_94e9e3b0',
    fileTypeIcon: 'fileTypeIcon_94e9e3b0',
    fileTypeIconIcon: 'fileTypeIconIcon_94e9e3b0',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_94e9e3b0',
    iconColumnHeader: 'iconColumnHeader_94e9e3b0',
    fileRow: 'fileRow_94e9e3b0',
    folderRow: 'folderRow_94e9e3b0',
    emptyFolder: 'emptyFolder_94e9e3b0',
    emptyFolderImage: 'emptyFolderImage_94e9e3b0',
    emptyFolderImageTag: 'emptyFolderImageTag_94e9e3b0',
    emptyFolderTitle: 'emptyFolderTitle_94e9e3b0',
    emptyFolderSubText: 'emptyFolderSubText_94e9e3b0',
    emptyFolderMobile: 'emptyFolderMobile_94e9e3b0',
    emptyFolderPc: 'emptyFolderPc_94e9e3b0',
    emptyFolderIcon: 'emptyFolderIcon_94e9e3b0'
};
export default styles;
/* tslint:enable */
