// @microsoft/sp-application-base/lib-commonjs/index.js
"use strict";
/**
 * SharePoint Framework support for building and extending client-side applications.
 *
 * @packagedocumentation
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
// Application Customizer
var ApplicationCustomizerContext_1 = require("./extensibility/ApplicationCustomizerContext");
Object.defineProperty(exports, "ApplicationCustomizerContext", { enumerable: true, get: function () { return ApplicationCustomizerContext_1.default; } });
var BaseApplicationCustomizer_1 = require("./extensibility/BaseApplicationCustomizer");
Object.defineProperty(exports, "BaseApplicationCustomizer", { enumerable: true, get: function () { return BaseApplicationCustomizer_1.default; } });
var ApplicationAccessor_1 = require("./extensibility/ApplicationAccessor");
Object.defineProperty(exports, "ApplicationAccessor", { enumerable: true, get: function () { return ApplicationAccessor_1.default; } });
var DialogManagerConfiguration_1 = require("./extensibility/DialogManagerConfiguration");
Object.defineProperty(exports, "_DialogManagerConfiguration", { enumerable: true, get: function () { return DialogManagerConfiguration_1.default; } });
var SearchQueryManager_1 = require("./extensibility/SearchQueryManager");
Object.defineProperty(exports, "_SearchQueryManager", { enumerable: true, get: function () { return SearchQueryManager_1.default; } });
// Placeholders
var PlaceholderContent_1 = require("./extensibility/placeholder/PlaceholderContent");
Object.defineProperty(exports, "PlaceholderContent", { enumerable: true, get: function () { return PlaceholderContent_1.default; } });
var PlaceholderName_1 = require("./extensibility/placeholder/PlaceholderName");
Object.defineProperty(exports, "PlaceholderName", { enumerable: true, get: function () { return PlaceholderName_1.default; } });
var Placeholder_1 = require("./extensibility/placeholder/Placeholder");
Object.defineProperty(exports, "_Placeholder", { enumerable: true, get: function () { return Placeholder_1.default; } });
var PlaceholderProvider_1 = require("./extensibility/placeholder/PlaceholderProvider");
Object.defineProperty(exports, "PlaceholderProvider", { enumerable: true, get: function () { return PlaceholderProvider_1.default; } });
var PlaceholderManager_1 = require("./extensibility/placeholder/PlaceholderManager");
Object.defineProperty(exports, "_PlaceholderManager", { enumerable: true, get: function () { return PlaceholderManager_1.default; } });
// Base application
var BaseApplication_1 = require("./BaseApplication");
Object.defineProperty(exports, "BaseApplication", { enumerable: true, get: function () { return BaseApplication_1.default; } });
var OnBeforeNavigationEventName_1 = require("./navigator/OnBeforeNavigationEventName");
Object.defineProperty(exports, "_OnBeforeNavigationEventName", { enumerable: true, get: function () { return OnBeforeNavigationEventName_1.default; } });
var BaseApplicationContext_1 = require("./BaseApplicationContext");
Object.defineProperty(exports, "BaseApplicationContext", { enumerable: true, get: function () { return BaseApplicationContext_1.default; } });
var ApplicationManager_1 = require("./ApplicationManager");
Object.defineProperty(exports, "_ApplicationManager", { enumerable: true, get: function () { return ApplicationManager_1.default; } });
var Navigator_1 = require("./navigator/Navigator");
Object.defineProperty(exports, "_Navigator", { enumerable: true, get: function () { return Navigator_1.default; } });
var NavigationDataError_1 = require("./navigator/NavigationDataError");
Object.defineProperty(exports, "_NavigationDataError", { enumerable: true, get: function () { return NavigationDataError_1.default; } });
Object.defineProperty(exports, "_NavigationErrorCode", { enumerable: true, get: function () { return NavigationDataError_1.NavigationErrorCode; } });
var PrefetchedDataEventArgs_1 = require("./navigator/PrefetchedDataEventArgs");
Object.defineProperty(exports, "_PrefetchedDataEventArgs", { enumerable: true, get: function () { return PrefetchedDataEventArgs_1.default; } });
var ApplicationLoadType_1 = require("./ApplicationLoadType");
Object.defineProperty(exports, "_ApplicationLoadType", { enumerable: true, get: function () { return ApplicationLoadType_1.ApplicationLoadType; } });
// Page Chrome
var SPThemeProvider_1 = require("./pageChrome/SPThemeProvider");
Object.defineProperty(exports, "_SPThemeProvider", { enumerable: true, get: function () { return SPThemeProvider_1.default; } });
/*
@internal
*/
tslib_1.__exportStar(require("@ms/odsp-datasources/lib/providers/theming/WebThemeProvider"), exports);
/*
@internal
*/
tslib_1.__exportStar(require("@ms/odsp-datasources/lib/providers/theming/ThemeUtilities"), exports);
// export * from '@ms/odsp-datasources/lib/providers/theming/ThemeInitializer';
// Unfortunately, themeInitializer causes errors if it is exported.
/*
@internal
*/
tslib_1.__exportStar(require("@ms/odsp-datasources/lib/dataSources/theming/spList/WebThemeDataSource"), exports);
/*
@internal
*/
var WebThemeDataSource_1 = require("@ms/odsp-datasources/lib/dataSources/theming/spList/WebThemeDataSource");
Object.defineProperty(exports, "WebThemeDataSource", { enumerable: true, get: function () { return WebThemeDataSource_1.default; } });
