// @rushstack/heft-config-file/lib/ConfigurationFile.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigurationFile = exports.PathResolutionMethod = exports.InheritanceType = void 0;
const nodeJsPath = __importStar(require("path"));
const jsonpath_plus_1 = require("jsonpath-plus");
const node_core_library_1 = require("@rushstack/node-core-library");
/**
 * @beta
 */
var InheritanceType;
(function (InheritanceType) {
    /**
     * Append additional elements after elements from the parent file's property
     */
    InheritanceType["append"] = "append";
    /**
     * Discard elements from the parent file's property
     */
    InheritanceType["replace"] = "replace";
    /**
     * Custom inheritance functionality
     */
    InheritanceType["custom"] = "custom";
})(InheritanceType = exports.InheritanceType || (exports.InheritanceType = {}));
/**
 * @beta
 */
var PathResolutionMethod;
(function (PathResolutionMethod) {
    /**
     * Resolve a path relative to the configuration file
     */
    PathResolutionMethod[PathResolutionMethod["resolvePathRelativeToConfigurationFile"] = 0] = "resolvePathRelativeToConfigurationFile";
    /**
     * Resolve a path relative to the root of the project containing the configuration file
     */
    PathResolutionMethod[PathResolutionMethod["resolvePathRelativeToProjectRoot"] = 1] = "resolvePathRelativeToProjectRoot";
    /**
     * Treat the property as a NodeJS-style require/import reference and resolve using standard
     * NodeJS filesystem resolution
     */
    PathResolutionMethod[PathResolutionMethod["NodeResolve"] = 2] = "NodeResolve";
})(PathResolutionMethod = exports.PathResolutionMethod || (exports.PathResolutionMethod = {}));
const CONFIGURATION_FILE_FIELD_ANNOTATION = Symbol('configuration-file-field-annotation');
/**
 * @beta
 */
class ConfigurationFile {
    constructor(options) {
        this._configurationFileCache = new Map();
        this._fileExistsCache = new Map();
        this._packageJsonLookup = new node_core_library_1.PackageJsonLookup();
        this._projectRelativeFilePath = options.projectRelativeFilePath;
        this._schemaPath = options.jsonSchemaPath;
        this._jsonPathMetadata = options.jsonPathMetadata || {};
        this._propertyInheritanceTypes = options.propertyInheritance || {};
    }
    get _schema() {
        if (!this.__schema) {
            this.__schema = node_core_library_1.JsonSchema.fromFile(this._schemaPath);
        }
        return this.__schema;
    }
    async loadConfigurationFileForProjectAsync(terminal, projectPath, rigConfig) {
        const projectConfigurationFilePath = this._getConfigurationFilePathForProject(projectPath);
        return await this._loadConfigurationFileInnerWithCacheAsync(terminal, projectConfigurationFilePath, new Set(), rigConfig);
    }
    /**
     * This function is identical to {@link ConfigurationFile.loadConfigurationFileForProjectAsync}, except
     * that a preliminary file existence check is performed and this function returns `undefined` if the
     * configuration file doesn't exist.
     */
    async tryLoadConfigurationFileForProjectAsync(terminal, projectPath, rigConfig) {
        const projectConfigurationFilePath = this._getConfigurationFilePathForProject(projectPath);
        const projectConfigurationFilePathForLogging = ConfigurationFile._formatPathForLogging(projectConfigurationFilePath);
        let exists = this._fileExistsCache.get(projectConfigurationFilePath);
        if (exists === undefined) {
            exists = await node_core_library_1.FileSystem.existsAsync(projectConfigurationFilePath);
            this._fileExistsCache.set(projectConfigurationFilePath, exists);
        }
        if (!exists) {
            if (rigConfig) {
                terminal.writeVerboseLine(`Config file "${projectConfigurationFilePathForLogging}" does not exist. Attempting to load via rig.`);
                return await this._tryLoadConfigurationFileInRigAsync(terminal, rigConfig, new Set());
            }
            else {
                return undefined;
            }
        }
        else {
            return await this._loadConfigurationFileInnerWithCacheAsync(terminal, projectConfigurationFilePath, new Set(), undefined);
        }
    }
    /**
     * Get the path to the source file that the referenced property was originally
     * loaded from.
     */
    getObjectSourceFilePath(obj) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const annotation = obj[CONFIGURATION_FILE_FIELD_ANNOTATION];
        if (annotation) {
            return annotation.configurationFilePath;
        }
        return undefined;
    }
    /**
     * Get the value of the specified property on the specified object that was originally
     * loaded from a configuration file.
     */
    getPropertyOriginalValue(options) {
        const annotation = 
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        options.parentObject[CONFIGURATION_FILE_FIELD_ANNOTATION];
        if (annotation && annotation.originalValues.hasOwnProperty(options.propertyName)) {
            return annotation.originalValues[options.propertyName];
        }
        else {
            return undefined;
        }
    }
    async _loadConfigurationFileInnerWithCacheAsync(terminal, resolvedConfigurationFilePath, visitedConfigurationFilePaths, rigConfig) {
        let cacheEntry = this._configurationFileCache.get(resolvedConfigurationFilePath);
        if (!cacheEntry) {
            try {
                cacheEntry = {
                    configurationFile: await this._loadConfigurationFileInnerAsync(terminal, resolvedConfigurationFilePath, visitedConfigurationFilePaths, rigConfig)
                };
            }
            catch (e) {
                cacheEntry = { error: e };
            }
        }
        else {
            terminal.writeVerboseLine(`Found "${ConfigurationFile._formatPathForLogging(resolvedConfigurationFilePath)}" in ConfigurationFile path.`);
        }
        if (cacheEntry.error) {
            throw cacheEntry.error;
        }
        else {
            return cacheEntry.configurationFile;
        }
    }
    async _loadConfigurationFileInnerAsync(terminal, resolvedConfigurationFilePath, visitedConfigurationFilePaths, rigConfig) {
        const resolvedConfigurationFilePathForLogging = ConfigurationFile._formatPathForLogging(resolvedConfigurationFilePath);
        if (visitedConfigurationFilePaths.has(resolvedConfigurationFilePath)) {
            throw new Error('A loop has been detected in the "extends" properties of configuration file at ' +
                `"${resolvedConfigurationFilePathForLogging}".`);
        }
        visitedConfigurationFilePaths.add(resolvedConfigurationFilePath);
        let fileText;
        try {
            fileText = await node_core_library_1.FileSystem.readFileAsync(resolvedConfigurationFilePath);
        }
        catch (e) {
            if (node_core_library_1.FileSystem.isNotExistError(e)) {
                terminal.writeVerboseLine(`Configuration file "${resolvedConfigurationFilePathForLogging}" not found.`);
                if (rigConfig) {
                    const rigResult = await this._tryLoadConfigurationFileInRigAsync(terminal, rigConfig, visitedConfigurationFilePaths);
                    if (rigResult) {
                        return rigResult;
                    }
                }
                e.message = `File does not exist: ${resolvedConfigurationFilePathForLogging}`;
            }
            throw e;
        }
        let configurationJson;
        try {
            configurationJson = await node_core_library_1.JsonFile.parseString(fileText);
        }
        catch (e) {
            throw new Error(`In config file "${resolvedConfigurationFilePathForLogging}": ${e}`);
        }
        this._schema.validateObject(configurationJson, resolvedConfigurationFilePathForLogging);
        this._annotateProperties(resolvedConfigurationFilePath, configurationJson);
        for (const [jsonPath, metadata] of Object.entries(this._jsonPathMetadata)) {
            jsonpath_plus_1.JSONPath({
                path: jsonPath,
                json: configurationJson,
                callback: (payload, payloadType, fullPayload) => {
                    if (metadata.pathResolutionMethod !== undefined) {
                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        fullPayload.parent[fullPayload.parentProperty] = this._resolvePathProperty(resolvedConfigurationFilePath, fullPayload.value, metadata.pathResolutionMethod);
                    }
                },
                otherTypeCallback: () => {
                    throw new Error('@other() tags are not supported');
                }
            });
        }
        let parentConfiguration = {};
        if (configurationJson.extends) {
            try {
                const resolvedParentConfigPath = node_core_library_1.Import.resolveModule({
                    modulePath: configurationJson.extends,
                    baseFolderPath: nodeJsPath.dirname(resolvedConfigurationFilePath)
                });
                parentConfiguration = await this._loadConfigurationFileInnerWithCacheAsync(terminal, resolvedParentConfigPath, visitedConfigurationFilePaths, undefined);
            }
            catch (e) {
                if (node_core_library_1.FileSystem.isNotExistError(e)) {
                    throw new Error(`In file "${resolvedConfigurationFilePathForLogging}", file referenced in "extends" property ` +
                        `("${configurationJson.extends}") cannot be resolved.`);
                }
                else {
                    throw e;
                }
            }
        }
        const propertyNames = new Set([
            ...Object.keys(parentConfiguration),
            ...Object.keys(configurationJson)
        ]);
        const resultAnnotation = {
            configurationFilePath: resolvedConfigurationFilePath,
            originalValues: {}
        };
        const result = {
            [CONFIGURATION_FILE_FIELD_ANNOTATION]: resultAnnotation
        };
        for (const propertyName of propertyNames) {
            if (propertyName === '$schema' || propertyName === 'extends') {
                continue;
            }
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const propertyValue = configurationJson[propertyName];
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const parentPropertyValue = parentConfiguration[propertyName];
            const bothAreArrays = Array.isArray(propertyValue) && Array.isArray(parentPropertyValue);
            const defaultInheritanceType = bothAreArrays
                ? { inheritanceType: InheritanceType.append }
                : { inheritanceType: InheritanceType.replace };
            const propertyInheritance = 
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            this._propertyInheritanceTypes[propertyName] !== undefined
                ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
                    this._propertyInheritanceTypes[propertyName]
                : defaultInheritanceType;
            let newValue;
            const usePropertyValue = () => {
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                resultAnnotation.originalValues[propertyName] = this.getPropertyOriginalValue({
                    parentObject: configurationJson,
                    propertyName: propertyName
                });
                newValue = propertyValue;
            };
            const useParentPropertyValue = () => {
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                resultAnnotation.originalValues[propertyName] = this.getPropertyOriginalValue({
                    parentObject: parentConfiguration,
                    propertyName: propertyName
                });
                newValue = parentPropertyValue;
            };
            if (propertyValue && !parentPropertyValue) {
                usePropertyValue();
            }
            else if (parentPropertyValue && !propertyValue) {
                useParentPropertyValue();
            }
            else {
                switch (propertyInheritance.inheritanceType) {
                    case InheritanceType.replace: {
                        if (propertyValue !== undefined) {
                            usePropertyValue();
                        }
                        else {
                            useParentPropertyValue();
                        }
                        break;
                    }
                    case InheritanceType.append: {
                        if (propertyValue !== undefined && parentPropertyValue === undefined) {
                            usePropertyValue();
                        }
                        else if (propertyValue === undefined && parentPropertyValue !== undefined) {
                            useParentPropertyValue();
                        }
                        else {
                            if (!Array.isArray(propertyValue) || !Array.isArray(parentPropertyValue)) {
                                throw new Error(`Issue in processing configuration file property "${propertyName}". ` +
                                    `Property is not an array, but the inheritance type is set as "${InheritanceType.append}"`);
                            }
                            newValue = [...parentPropertyValue, ...propertyValue];
                            newValue[CONFIGURATION_FILE_FIELD_ANNOTATION] = {
                                configurationFilePath: undefined,
                                originalValues: Object.assign(Object.assign({}, parentPropertyValue[CONFIGURATION_FILE_FIELD_ANNOTATION].originalValues), propertyValue[CONFIGURATION_FILE_FIELD_ANNOTATION].originalValues)
                            };
                        }
                        break;
                    }
                    case InheritanceType.custom: {
                        const customInheritance = propertyInheritance;
                        if (!customInheritance.inheritanceFunction ||
                            typeof customInheritance.inheritanceFunction !== 'function') {
                            throw new Error('For property inheritance type "InheritanceType.custom", an inheritanceFunction must be provided.');
                        }
                        newValue = customInheritance.inheritanceFunction(propertyValue, parentPropertyValue);
                        break;
                    }
                    default: {
                        throw new Error(`Unknown inheritance type "${propertyInheritance}"`);
                    }
                }
            }
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            result[propertyName] = newValue;
        }
        try {
            this._schema.validateObject(result, resolvedConfigurationFilePathForLogging);
        }
        catch (e) {
            throw new Error(`Resolved configuration object does not match schema: ${e}`);
        }
        return result;
    }
    async _tryLoadConfigurationFileInRigAsync(terminal, rigConfig, visitedConfigurationFilePaths) {
        if (rigConfig.rigFound) {
            const rigProfileFolder = await rigConfig.getResolvedProfileFolderAsync();
            try {
                return await this._loadConfigurationFileInnerWithCacheAsync(terminal, nodeJsPath.resolve(rigProfileFolder, this._projectRelativeFilePath), visitedConfigurationFilePaths, undefined);
            }
            catch (e) {
                // Ignore cases where a configuration file doesn't exist in a rig
                if (!node_core_library_1.FileSystem.isNotExistError(e)) {
                    throw e;
                }
                else {
                    terminal.writeVerboseLine(`Configuration file "${this._projectRelativeFilePath}" not found in rig ("${ConfigurationFile._formatPathForLogging(rigProfileFolder)}")`);
                }
            }
        }
        else {
            terminal.writeVerboseLine(`No rig found for "${ConfigurationFile._formatPathForLogging(rigConfig.projectFolderPath)}"`);
        }
        return undefined;
    }
    _annotateProperties(resolvedConfigurationFilePath, obj) {
        if (!obj) {
            return;
        }
        if (typeof obj === 'object') {
            this._annotateProperty(resolvedConfigurationFilePath, obj);
            for (const objValue of Object.values(obj)) {
                this._annotateProperties(resolvedConfigurationFilePath, objValue);
            }
        }
    }
    _annotateProperty(resolvedConfigurationFilePath, obj) {
        if (!obj) {
            return;
        }
        if (typeof obj === 'object') {
            obj[CONFIGURATION_FILE_FIELD_ANNOTATION] = {
                configurationFilePath: resolvedConfigurationFilePath,
                originalValues: Object.assign({}, obj)
            };
        }
    }
    _resolvePathProperty(configurationFilePath, propertyValue, resolutionMethod) {
        switch (resolutionMethod) {
            case PathResolutionMethod.resolvePathRelativeToConfigurationFile: {
                return nodeJsPath.resolve(nodeJsPath.dirname(configurationFilePath), propertyValue);
            }
            case PathResolutionMethod.resolvePathRelativeToProjectRoot: {
                const packageRoot = this._packageJsonLookup.tryGetPackageFolderFor(configurationFilePath);
                if (!packageRoot) {
                    throw new Error(`Could not find a package root for path "${ConfigurationFile._formatPathForLogging(configurationFilePath)}"`);
                }
                return nodeJsPath.resolve(packageRoot, propertyValue);
            }
            case PathResolutionMethod.NodeResolve: {
                return node_core_library_1.Import.resolveModule({
                    modulePath: propertyValue,
                    baseFolderPath: nodeJsPath.dirname(configurationFilePath)
                });
            }
            default: {
                return propertyValue;
            }
        }
    }
    _getConfigurationFilePathForProject(projectPath) {
        return nodeJsPath.resolve(projectPath, this._projectRelativeFilePath);
    }
}
exports.ConfigurationFile = ConfigurationFile;
/**
 * @internal
 */
ConfigurationFile._formatPathForLogging = (path) => path;
