// @microsoft/sp-core-library/lib-commonjs/index.js
"use strict";
/**
 * SharePoint Framework core libraries.
 *
 * @remarks
 * This package provides a foundation of core classes that ensure a consistent character
 * across all the other SharePoint Framework APIs.  Because this package is a dependency
 * of every other package, its design prioritizes small code size and broad applicability,
 * versus flexibility or richness of features.
 *
 * @packagedocumentation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventArgs = exports.Event = void 0;
var tslib_1 = require("tslib");
var BrowserDetection_1 = require("./BrowserDetection");
Object.defineProperty(exports, "_BrowserDetection", { enumerable: true, get: function () { return BrowserDetection_1.default; } });
Object.defineProperty(exports, "_Browser", { enumerable: true, get: function () { return BrowserDetection_1.Browser; } });
Object.defineProperty(exports, "_OS", { enumerable: true, get: function () { return BrowserDetection_1.OS; } });
var DisplayMode_1 = require("./DisplayMode");
Object.defineProperty(exports, "DisplayMode", { enumerable: true, get: function () { return DisplayMode_1.default; } });
var ClientManifestData_1 = require("./ClientManifestData");
Object.defineProperty(exports, "_ClientManifestData", { enumerable: true, get: function () { return ClientManifestData_1.default; } });
var Guid_1 = require("./Guid");
Object.defineProperty(exports, "Guid", { enumerable: true, get: function () { return Guid_1.default; } });
var Text_1 = require("./Text");
Object.defineProperty(exports, "Text", { enumerable: true, get: function () { return Text_1.default; } });
var Version_1 = require("./Version");
Object.defineProperty(exports, "Version", { enumerable: true, get: function () { return Version_1.default; } });
var Validate_1 = require("./Validate");
Object.defineProperty(exports, "Validate", { enumerable: true, get: function () { return Validate_1.default; } });
var UrlUtilities_1 = require("./url/UrlUtilities");
Object.defineProperty(exports, "UrlUtilities", { enumerable: true, get: function () { return UrlUtilities_1.default; } });
var UrlQueryParameterCollection_1 = require("./url/UrlQueryParameterCollection");
Object.defineProperty(exports, "UrlQueryParameterCollection", { enumerable: true, get: function () { return UrlQueryParameterCollection_1.default; } });
var TimeProvider_1 = require("./serviceScope/provider/TimeProvider");
Object.defineProperty(exports, "TimeProvider", { enumerable: true, get: function () { return TimeProvider_1.default; } });
var RandomNumberGenerator_1 = require("./serviceScope/provider/RandomNumberGenerator");
Object.defineProperty(exports, "RandomNumberGenerator", { enumerable: true, get: function () { return RandomNumberGenerator_1.default; } });
var ServiceKey_1 = require("./serviceScope/ServiceKey");
Object.defineProperty(exports, "ServiceKey", { enumerable: true, get: function () { return ServiceKey_1.default; } });
var ServiceScope_1 = require("./serviceScope/ServiceScope");
Object.defineProperty(exports, "ServiceScope", { enumerable: true, get: function () { return ServiceScope_1.default; } });
var Session_1 = require("./Session");
Object.defineProperty(exports, "Session", { enumerable: true, get: function () { return Session_1.default; } });
var Environment_1 = require("./Environment");
Object.defineProperty(exports, "Environment", { enumerable: true, get: function () { return Environment_1.default; } });
Object.defineProperty(exports, "EnvironmentType", { enumerable: true, get: function () { return Environment_1.EnvironmentType; } });
var SPFlight_1 = require("./SPFlight");
Object.defineProperty(exports, "_SPFlight", { enumerable: true, get: function () { return SPFlight_1.default; } });
var SPKillSwitch_1 = require("./SPKillSwitch");
Object.defineProperty(exports, "_SPKillSwitch", { enumerable: true, get: function () { return SPKillSwitch_1.default; } });
var SPExperiment_1 = require("./SPExperiment");
Object.defineProperty(exports, "_SPExperiment", { enumerable: true, get: function () { return SPExperiment_1.SPExperiment; } });
var SPError_1 = require("./SPError");
Object.defineProperty(exports, "SPError", { enumerable: true, get: function () { return SPError_1.default; } });
var Log_1 = require("./log/Log");
Object.defineProperty(exports, "Log", { enumerable: true, get: function () { return Log_1.default; } });
// Events
var SPEvent_1 = require("./events/SPEvent");
Object.defineProperty(exports, "SPEvent", { enumerable: true, get: function () { return SPEvent_1.default; } });
var SPEventArgs_1 = require("./events/SPEventArgs");
Object.defineProperty(exports, "SPEventArgs", { enumerable: true, get: function () { return SPEventArgs_1.default; } });
var SPEventManager_1 = require("./events/SPEventManager");
Object.defineProperty(exports, "_SPEventManager", { enumerable: true, get: function () { return SPEventManager_1.default; } });
// Backwards compatibility for bundles that referenced the originally shipped names
// "Event" and "EventArgs" (which turned out to conflict with the DOM API).
var SPEvent_2 = tslib_1.__importDefault(require("./events/SPEvent"));
var SPEventArgs_2 = tslib_1.__importDefault(require("./events/SPEventArgs"));
/**
 * DO NOT USE THIS DEFINITION.
 * @remarks
 * This is a backwards compatibility shim for legacy third-party bundles that imported the
 * old "Event" class.  It was renamed to solve a naming conflict with the brower DOM "Event".
 * @internal @preapproved
 * @deprecated Use the SPEvent class instead.
 */
var Event = /** @class */ (function (_super) {
    tslib_1.__extends(Event, _super);
    function Event() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Event;
}(SPEvent_2.default));
exports.Event = Event;
/**
 * DO NOT USE THIS DEFINITION.
 * @remarks
 * This is a backwards compatibility shim for legacy third-party bundles that imported the
 * old "EventArgs" class.  It was renamed to solve a naming conflict with the brower DOM "Event".
 * @internal @preapproved
 * @deprecated Use the SPEvent class instead.
 */
var EventArgs = /** @class */ (function (_super) {
    tslib_1.__extends(EventArgs, _super);
    function EventArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return EventArgs;
}(SPEventArgs_2.default));
exports.EventArgs = EventArgs;
var DeferredClass_1 = require("./deferredClass/DeferredClass");
Object.defineProperty(exports, "_DeferredClass", { enumerable: true, get: function () { return DeferredClass_1.default; } });
var BrowserUtilities_1 = require("./BrowserUtilities");
Object.defineProperty(exports, "_BrowserUtilities", { enumerable: true, get: function () { return BrowserUtilities_1.BrowserUtilities; } });
var Envoy_1 = require("./envoy/Envoy");
Object.defineProperty(exports, "_Envoy", { enumerable: true, get: function () { return Envoy_1.Envoy; } });
var ServiceKeys_1 = require("./serviceScope/serviceKeys/ServiceKeys");
Object.defineProperty(exports, "_ServiceKeys", { enumerable: true, get: function () { return ServiceKeys_1.ServiceKeys; } });
var WebWorkerProxy_1 = require("./webWorker/WebWorkerProxy");
Object.defineProperty(exports, "_WebWorkerProxy", { enumerable: true, get: function () { return WebWorkerProxy_1.WebWorkerProxy; } });
