// @microsoft/rush-lib/lib/cli/actions/UpdateAutoinstallerAction.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateAutoinstallerAction = void 0;
const BaseRushAction_1 = require("./BaseRushAction");
const Autoinstaller_1 = require("../../logic/Autoinstaller");
class UpdateAutoinstallerAction extends BaseRushAction_1.BaseRushAction {
    constructor(parser) {
        super({
            actionName: 'update-autoinstaller',
            summary: 'Updates autoinstaller package dependenices',
            documentation: 'Use this command to regenerate the shrinkwrap file for an autoinstaller folder.',
            parser
        });
    }
    onDefineParameters() {
        this._name = this.defineStringParameter({
            parameterLongName: '--name',
            argumentName: 'AUTOINSTALLER_NAME',
            required: true,
            description: 'Specifies the name of the autoinstaller, which must be one of the folders under common/autoinstallers.'
        });
    }
    async runAsync() {
        const autoinstallerName = this._name.value;
        const autoinstaller = new Autoinstaller_1.Autoinstaller(autoinstallerName, this.rushConfiguration);
        autoinstaller.update();
        console.log('\nSuccess.');
    }
}
exports.UpdateAutoinstallerAction = UpdateAutoinstallerAction;
