// @microsoft/sp-diagnostics/lib/Api/Qos/QosMonitor.js
// Copyright (c) Microsoft. All rights reserved.
import { Guid, Text, Validate } from '@microsoft/sp-core-library';
import TraceLogger from '../Trace/TraceLogger';
import { QosResultType } from './QosResultType';
import QosLogger from './QosLogger';
import EngagementLogger from '../Engagement/EngagementLogger';
import DiagnosticsSettingsManager from '../../DiagnosticsSettingsManager';
/**
 * Qos logger
 * This class represents the Qos monitor APIs
 *
 * @preapproved @internal
 */
var QosMonitor = /** @class */ (function () {
    /**
     * Creates a QoS monitor (initializes it and logs the start event)
     * @param scenarioName	- Unique name of the QoS scenario
     * You will see start tag: SPPages.ScenarioName.Start
     * @param copyToEngagement - indicate whether this data should be used as Engagement as well.
     */
    function QosMonitor(scenarioName, copyToEngagement) {
        var _a;
        if (copyToEngagement === void 0) { copyToEngagement = false; }
        this._qosResultTypeValue = (_a = {},
            _a[QosResultType.Success] = 'Success',
            _a[QosResultType.Failure] = 'Failure',
            _a[QosResultType.ExpectedFailure] = 'ExpectedFailure',
            _a);
        Validate.isNonemptyString(scenarioName, 'scenarioName');
        this._id = Guid.newGuid().toString();
        if (!this._isDebugSession()) {
            QosLogger.instance.startQosMonitor(this._id, { name: scenarioName });
        }
        this._scenarioName = scenarioName;
        this._hasEnded = false;
        this._copyToEngagement = copyToEngagement;
    }
    QosMonitor.startRealTimeProcessing = function () {
        if (!QosLogger.instance.shouldProcessInRealTime) {
            this._shouldProcessInRealTime(true);
        }
    };
    /**
     * Only used in unit tests
     * @internal
     */
    QosMonitor._shouldProcessInRealTime = function (processInRealTime) {
        QosLogger.instance.shouldProcessInRealTime = processInRealTime;
    };
    Object.defineProperty(QosMonitor.prototype, "name", {
        get: function () {
            return this._scenarioName;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(QosMonitor.prototype, "shortEventName", {
        get: function () {
            return 'Qos';
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(QosMonitor.prototype, "resultCode", {
        get: function () {
            return this._endResultCode;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(QosMonitor.prototype, "resultType", {
        get: function () {
            return this._endResult;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(QosMonitor.prototype, "extraData", {
        get: function () {
            return this._endExtraData;
        },
        set: function (value) {
            this._endExtraData = value;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(QosMonitor.prototype, "hasEnded", {
        get: function () {
            return this._hasEnded;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * QoS monitor ends with success
     * You will see success tag: SPPages.ScenarioName.Success
     */
    // tslint:disable-next-line:no-any
    QosMonitor.prototype.writeSuccess = function (extraData) {
        this._writeQosEnd(QosResultType.Success, undefined, undefined, extraData);
    };
    /**
     * QoS monitor ends with failure
     * You will see failure tag: SPPages.ScenarioName.<failuretag>.Failure.
     * <failuretag> will be filled with param tagName
     */
    QosMonitor.prototype.writeUnexpectedFailure = function (tagNameSuffix, ex, extraData // tslint:disable-line:no-any
    ) {
        if (tagNameSuffix === void 0) { tagNameSuffix = 'DefaultUnexpected'; }
        this._writeQosEnd(QosResultType.Failure, tagNameSuffix, ex, extraData);
    };
    /*
     * QoS monitor ends with expected failure
     * You will see success tag: ModernPublish.RenderPublishPage.<failuretag>.ExpectedFailure.
     * <failuretag> will be filled with param tagName
     */
    QosMonitor.prototype.writeExpectedFailure = function (tagNameSuffix, ex, extraData // tslint:disable-line:no-any
    ) {
        if (tagNameSuffix === void 0) { tagNameSuffix = 'DefaultExpected'; }
        this._writeQosEnd(QosResultType.ExpectedFailure, tagNameSuffix, ex, extraData);
    };
    QosMonitor.prototype._writeQosEnd = function (resultType, tagNameSuffix, ex, extraData // tslint:disable-line:no-any
    ) {
        if (this._shouldSkip(resultType)) {
            return;
        }
        this._ensureCorrelationId(resultType, extraData);
        this._endResultCode = tagNameSuffix;
        this._endResult = resultType;
        var endData = {
            resultCode: this._endResultCode,
            resultType: this._endResult,
            error: ex ? ex.message : undefined,
            extraData: this._endExtraData
        };
        QosLogger.instance.writeQosEndResult(this._id, endData);
        this._writeToEngagement();
    };
    QosMonitor.prototype._ensureCorrelationId = function (resultType, extraData // tslint:disable-line:no-any
    ) {
        var _this = this;
        if (extraData) {
            if (!this._endExtraData) {
                this._endExtraData = {};
            }
            // Instead of slower ...[spread], copy over the incoming values manually
            Object.keys(extraData).forEach(function (k) { return (_this._endExtraData[k] = extraData[k]); });
        }
        if (DiagnosticsSettingsManager.isInitialized && resultType !== QosResultType.Success) {
            var capitalKey = 'CorrelationId';
            var lowerKey = 'correlationId';
            var correlationId = (this._endExtraData && (this._endExtraData[capitalKey] || this._endExtraData[lowerKey])) ||
                DiagnosticsSettingsManager.settings.correlationId;
            if (correlationId) {
                if (!this._endExtraData) {
                    this._endExtraData = {};
                }
                this._endExtraData[capitalKey] = correlationId;
            }
        }
    };
    QosMonitor.prototype._isDebugSession = function () {
        return !!sessionStorage.getItem('spfx-debug');
    };
    QosMonitor.prototype._shouldSkip = function (result) {
        if (this._hasEnded && this._endResult !== undefined) {
            var message = Text.format(QosMonitor.qosMonitorHasEndedErrorString, this.name, this._qosResultTypeValue[this._endResult], this._qosResultTypeValue[result]);
            if (this._endResult !== result) {
                TraceLogger.logError(QosMonitor.logSource, new Error(message));
            }
            if (DEBUG) {
                TraceLogger.logVerbose(QosMonitor.logSource, message, 'skipEnd');
            }
            return true;
        }
        else {
            this._hasEnded = true;
            return this._isDebugSession();
        }
    };
    QosMonitor.prototype._writeToEngagement = function () {
        if (this._shouldWriteToEngagement()) {
            // Engagement name requires certain format. Cannot just use QOS name.
            // Record Qos name in extra data 'qosName'
            var engagementData = {
                name: 'Spfx.Qos.ToEngagement',
                extraData: this.extraData || {}
            };
            // tslint:disable-next-line:no-string-literal
            engagementData.extraData['qosName'] = this.name;
            if (this.resultType) {
                // tslint:disable-next-line:no-string-literal
                engagementData.extraData['qosResultType'] = this._qosResultTypeValue[this.resultType];
            }
            if (this.resultCode) {
                // tslint:disable-next-line:no-string-literal
                engagementData.extraData['qosResultCode'] = this.resultCode;
            }
            EngagementLogger.log(engagementData);
        }
    };
    QosMonitor.prototype._shouldWriteToEngagement = function () {
        var shouldWriteEngagement = this._copyToEngagement;
        return shouldWriteEngagement;
    };
    QosMonitor.logSource = {
        id: 'QosMonitor'
    };
    QosMonitor.qosMonitorHasEndedErrorString = "QoS monitor '{0}' has ended with '{1}' already. Ignoring '{2}' tag.";
    return QosMonitor;
}());
export default QosMonitor;
