// @microsoft/rush-lib/lib/api/LastInstallFlag.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LastInstallFlagFactory = exports.LastInstallFlag = exports.LAST_INSTALL_FLAG_FILE_NAME = void 0;
const path = __importStar(require("path"));
const node_core_library_1 = require("@rushstack/node-core-library");
const lodash = node_core_library_1.Import.lazy('lodash', require);
exports.LAST_INSTALL_FLAG_FILE_NAME = 'last-install.flag';
/**
 * A helper class for managing last-install flags, which are persistent and
 * indicate that something installed in the folder was successfully completed.
 * It also compares state, so that if something like the Node.js version has changed,
 * it can invalidate the last install.
 * @internal
 */
class LastInstallFlag {
    /**
     * Creates a new LastInstall flag
     * @param folderPath - the folder that this flag is managing
     * @param state - optional, the state that should be managed or compared
     */
    constructor(folderPath, state = {}) {
        this._path = path.join(folderPath, this.flagName);
        this._state = state;
    }
    /**
     * Returns true if the file exists and the contents match the current state.
     */
    isValid() {
        return this._isValid(false);
    }
    /**
     * Same as isValid(), but with an additional check:  If the current state is not equal to the previous
     * state, and an the current state causes an error, then throw an exception with a friendly message.
     *
     * @internal
     */
    checkValidAndReportStoreIssues() {
        return this._isValid(true);
    }
    _isValid(checkValidAndReportStoreIssues) {
        let oldState;
        try {
            oldState = node_core_library_1.JsonFile.load(this._path);
        }
        catch (err) {
            return false;
        }
        const newState = this._state;
        if (!lodash.isEqual(oldState, newState)) {
            if (checkValidAndReportStoreIssues) {
                const pkgManager = newState.packageManager;
                if (pkgManager === 'pnpm') {
                    if (
                    // Only throw an error if the package manager hasn't changed from PNPM
                    oldState.packageManager === pkgManager &&
                        // Throw if the store path changed
                        oldState.storePath !== newState.storePath) {
                        const oldStorePath = oldState.storePath || '<global>';
                        const newStorePath = newState.storePath || '<global>';
                        throw new Error('Current PNPM store path does not match the last one used. This may cause inconsistency in your builds.\n\n' +
                            'If you wish to install with the new store path, please run "rush update --purge"\n\n' +
                            `Old Path: ${oldStorePath}\n` +
                            `New Path: ${newStorePath}`);
                    }
                }
            }
            return false;
        }
        return true;
    }
    /**
     * Writes the flag file to disk with the current state
     */
    create() {
        node_core_library_1.JsonFile.save(this._state, this._path, {
            ensureFolderExists: true
        });
    }
    /**
     * Removes the flag file
     */
    clear() {
        node_core_library_1.FileSystem.deleteFile(this._path);
    }
    /**
     * Returns the full path to the flag file
     */
    get path() {
        return this._path;
    }
    /**
     * Returns the name of the flag file
     */
    get flagName() {
        return exports.LAST_INSTALL_FLAG_FILE_NAME;
    }
}
exports.LastInstallFlag = LastInstallFlag;
/**
 * A helper class for LastInstallFlag
 *
 * @internal
 */
class LastInstallFlagFactory {
    /**
     * Gets the LastInstall flag and sets the current state. This state is used to compare
     * against the last-known-good state tracked by the LastInstall flag.
     * @param rushConfiguration - the configuration of the Rush repo to get the install
     * state from
     *
     * @internal
     */
    static getCommonTempFlag(rushConfiguration) {
        const currentState = {
            node: process.versions.node,
            packageManager: rushConfiguration.packageManager,
            packageManagerVersion: rushConfiguration.packageManagerToolVersion
        };
        if (currentState.packageManager === 'pnpm' && rushConfiguration.pnpmOptions) {
            currentState.storePath = rushConfiguration.pnpmOptions.pnpmStorePath;
            if (rushConfiguration.pnpmOptions.useWorkspaces) {
                currentState.workspaces = rushConfiguration.pnpmOptions.useWorkspaces;
            }
        }
        return new LastInstallFlag(rushConfiguration.commonTempFolder, currentState);
    }
}
exports.LastInstallFlagFactory = LastInstallFlagFactory;
