// @microsoft/sp-loader/lib/requirejs/SPRequireJsComponentLoader.js
import { __extends } from "tslib";
// Copyright (c) Microsoft. All rights reserved.
import { Validate } from '@microsoft/sp-core-library';
import { _QosMonitor, _TraceLogger } from '@microsoft/sp-diagnostics';
import ErrorBuilder from '../error/ErrorBuilder';
import SPLoaderError from '../error/SPLoaderError';
import { BaseComponentLoader } from '../loader/BaseComponentLoader';
import _loadComponent from '../loader/loadComponent';
import ComponentStore from '../stores/ComponentStore';
import ComponentOverrider from '../utilities/ComponentOverrider';
import LoadComponentExecutor from '../utilities/LoadComponentExecutor';
import * as telemetryConstants from '../utilities/telemetryConstants';
import { DeferredSystemJsFallbackLoader } from './DeferredSystemJsFallbackLoader';
import RequireJsLoader from './RequireJsLoader';
/**
 * The component loader for SPFx.
 * This implementation uses RequireJS as the internal module loader.
 *
 * @alpha
 */
var SPRequireJsComponentLoader = /** @class */ (function (_super) {
    __extends(SPRequireJsComponentLoader, _super);
    // Initialization
    /**
     * @internal
     */
    function SPRequireJsComponentLoader(serviceScope, useSecondaryCdn) {
        var _this = _super.call(this, serviceScope) || this;
        if (useSecondaryCdn) {
            _TraceLogger.logError(telemetryConstants.spRequireJsComponentLoader, new Error(telemetryConstants.usingSecondaryCdnEventName), telemetryConstants.usingSecondaryCdnEventName);
        }
        _this._requireJsLoader = serviceScope.consume(RequireJsLoader.serviceKey);
        _this._requireJsLoader.useSecondaryCdn = useSecondaryCdn;
        _this._loadComponentExecutor = new LoadComponentExecutor(_this._loadComponentWithExecutor.bind(_this));
        return _this;
    }
    // Public API
    SPRequireJsComponentLoader.prototype.loadScript = function (url, options) {
        var _a;
        try {
            Validate.isNonemptyString(url, 'url');
        }
        catch (error) {
            return Promise.reject(error);
        }
        var qosMonitor = new _QosMonitor(telemetryConstants.loadScriptQosScenarioName);
        var rQosMonitor = new _QosMonitor(telemetryConstants.loadScriptWithRequireQosScenarioName);
        if (options === undefined) {
            options = {};
        }
        if (typeof options === 'string') {
            var error = ErrorBuilder.buildLoadScriptWithStringError();
            qosMonitor.writeExpectedFailure('OptionsAsString', error);
            return Promise.reject(error);
        }
        if (options.globalExportsName) {
            this._requireJsLoader.requireConfig({
                shim: (_a = {},
                    _a[url.replace(/\.js$/, '')] = {
                        exports: options.globalExportsName
                    },
                    _a)
            });
        }
        return this._requireJsLoader
            .requireLoad(url, options.globalExportsName)
            .then(function (module) {
            qosMonitor.writeSuccess();
            rQosMonitor.writeSuccess();
            return module;
        })
            .catch(function (e) {
            qosMonitor.writeUnexpectedFailure('RequireLoad', e);
            rQosMonitor.writeUnexpectedFailure('RequireLoad', e);
            throw e;
        });
    };
    SPRequireJsComponentLoader.prototype.loadComponent = function (manifest) {
        var _this = this;
        var qosMonitor;
        var qosExtraData;
        if (!manifest.isInternal) {
            qosMonitor = new _QosMonitor(telemetryConstants.loadComponentQosScenarioName);
            qosExtraData = this._buildQosExtraData(manifest);
        }
        return this._loadComponentExecutor
            .loadComponent(manifest)
            .then(function (componentLoaded) {
            var component = componentLoaded.component, requireJsError = componentLoaded.requireJsError;
            if (requireJsError && qosExtraData) {
                qosExtraData.loader = 'systemjs-fallback';
            }
            if (qosMonitor) {
                qosMonitor.writeSuccess(qosExtraData);
            }
            return component;
        })
            .catch(function (requireJsError) {
            return _this._handleLoadComponentError(requireJsError, manifest, qosMonitor, qosExtraData);
        });
    };
    // Internal/private API
    SPRequireJsComponentLoader.prototype._overrideComponent = function (componentId, componentModule) {
        ComponentOverrider.overrideComponent(componentId, componentModule, this._serviceScope, RequireJsLoader.serviceKey);
    };
    SPRequireJsComponentLoader.prototype._unloadComponent = function (manifest) {
        if (ComponentStore.instance.tryGetComponent(manifest.id, manifest.version)) {
            ComponentStore.instance.deleteComponent(manifest.id, manifest.version);
            this._requireJsLoader.requireDelete(manifest);
        }
    };
    /**
     * This functions tries to load the component with RequireJS, and if the killswitch is not enabled and the load fails,
     * will set up the component to be loaded with SystemJS.
     */
    SPRequireJsComponentLoader.prototype._loadComponentWithExecutor = function (manifest) {
        var _this = this;
        var loadWithRequireMonitor = new _QosMonitor(telemetryConstants.loadComponentWithRequireQosScenarioName);
        return _loadComponent(manifest, this._requireJsLoader)
            .then(function (component) {
            loadWithRequireMonitor.writeSuccess();
            return {
                component: component
            };
        })
            .catch(function (requireJsError) {
            loadWithRequireMonitor.writeUnexpectedFailure('RequireLoad', requireJsError);
            var loadWithFallbackMonitor = new _QosMonitor(telemetryConstants.loadComponentWithFallbackSystemQosScenarioName);
            loadWithFallbackMonitor.extraData = { isInternal: !!manifest.isInternal };
            _TraceLogger.logVerbose(telemetryConstants.loadComponentLogSource, "RequireJS failed to load component \"" + manifest.id + "\". Trying again with SystemJS.");
            // Remove the component from the store to ensure that SystemJS doesn't assume it's already loaded
            ComponentStore.instance.deleteComponent(manifest.id, manifest.version);
            return _this._ensureFallback()
                .then(function () { return _this._systemJsFallbackLoader.loadComponent(manifest); })
                .then(function (component) {
                loadWithFallbackMonitor.writeSuccess();
                return {
                    component: component,
                    requireJsError: requireJsError
                };
            })
                .catch(function () {
                loadWithFallbackMonitor.writeUnexpectedFailure('FallbackLoad', requireJsError);
                // The systemjs error was already logged by the systemjs loader
                throw requireJsError;
            });
        });
    };
    SPRequireJsComponentLoader.prototype._ensureFallback = function () {
        var _this = this;
        if (this._systemJsFallbackLoader) {
            return Promise.resolve();
        }
        return DeferredSystemJsFallbackLoader.getInstance()
            .load()
            .then(function () {
            _this._systemJsFallbackLoader = DeferredSystemJsFallbackLoader.getInstance().create(_this._serviceScope);
            _this._loadComponentExecutor.setAlternativeExecutor(_this._systemJsFallbackLoader.executor);
            _this._systemJsFallbackLoader.executor.setAlternativeExecutor(_this._loadComponentExecutor);
        });
    };
    SPRequireJsComponentLoader.prototype._buildQosExtraData = function (manifest) {
        return {
            manifestId: manifest.id,
            version: manifest.version,
            alias: manifest.alias,
            isInternal: manifest.isInternal,
            isDebug: manifest._isDebug,
            loader: 'requirejs'
        };
    };
    SPRequireJsComponentLoader.prototype._handleLoadComponentError = function (error, manifest, qosMonitor, qosExtraData) {
        if (error instanceof SPLoaderError && error.isExpected && qosMonitor) {
            qosMonitor.writeExpectedFailure(undefined, error, qosExtraData);
        }
        else if (qosMonitor) {
            qosMonitor.writeUnexpectedFailure(undefined, error, qosExtraData);
        }
        // If it fails to load, the caller can then call loadComponent() again and not hit the cache
        ComponentStore.instance.deleteComponent(manifest.id, manifest.version);
        throw error;
    };
    return SPRequireJsComponentLoader;
}(BaseComponentLoader));
export default SPRequireJsComponentLoader;
