// @microsoft/spfx-heft-plugins/lib/spfxHeftPlugin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const DevDeployPlugin_1 = __importDefault(require("./plugins/devDeployPlugin/DevDeployPlugin"));
const LocalizationTypingsPlugin_1 = __importDefault(require("./plugins/localizationTypingsPlugin/LocalizationTypingsPlugin"));
const WebpackServePlugin_1 = __importDefault(require("./plugins/webpackServePlugin/WebpackServePlugin"));
const WebpackConfigurationPlugin_1 = __importDefault(require("./plugins/webpackConfigurationPlugin/WebpackConfigurationPlugin"));
const CustomizeWebpackConfigurationPlugin_1 = __importDefault(require("./plugins/customizeWebpackConfigurationPlugin/CustomizeWebpackConfigurationPlugin"));
/**
 * @public
 */
class SpfxHeftPlugin {
    constructor() {
        this.pluginName = "spfx-heft-plugin" /* SpfxHeftPlugin */;
    }
    get accessor() {
        if (!this._accessor) {
            this._accessor = {
                devDeploy: DevDeployPlugin_1.default.accessor,
                spfxWebpackConfiguration: WebpackConfigurationPlugin_1.default.accessor
            };
        }
        return this._accessor;
    }
    apply(heftSession, heftConfiguration) {
        LocalizationTypingsPlugin_1.default.apply(heftSession, heftConfiguration);
        DevDeployPlugin_1.default.apply(heftSession, heftConfiguration);
        WebpackConfigurationPlugin_1.default.apply(heftSession, heftConfiguration);
        WebpackServePlugin_1.default.apply(heftSession, heftConfiguration);
        CustomizeWebpackConfigurationPlugin_1.default.apply(heftSession, heftConfiguration);
    }
}
exports.default = new SpfxHeftPlugin(); // tslint:disable-line:export-name
