// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/FilePicker.module.scss.js
/* tslint:disable */
require("./FilePicker.module.css");
var styles = {
    filePicker: 'filePicker_936a8d11',
    tabsContainer: 'tabsContainer_936a8d11',
    tabContainer: 'tabContainer_936a8d11',
    nav: 'nav_936a8d11',
    tabHeaderContainer: 'tabHeaderContainer_936a8d11',
    breadcrumbNav: 'breadcrumbNav_936a8d11',
    breadcrumbNavItem: 'breadcrumbNavItem_936a8d11',
    tab: 'tab_936a8d11',
    tabOffset: 'tabOffset_936a8d11',
    tabFiles: 'tabFiles_936a8d11',
    scrollablePaneWrapper: 'scrollablePaneWrapper_936a8d11',
    actionButtonsContainer: 'actionButtonsContainer_936a8d11',
    actionButtons: 'actionButtons_936a8d11',
    actionButton: 'actionButton_936a8d11',
    tabHeader: 'tabHeader_936a8d11',
    selectedTab: 'selectedTab_936a8d11',
    header: 'header_936a8d11',
    xlg: 'xlg_936a8d11',
    xxlg: 'xxlg_936a8d11',
    xxxlg: 'xxxlg_936a8d11',
    xxxxlg: 'xxxxlg_936a8d11',
    itemPickerTopBar: 'itemPickerTopBar_936a8d11',
    commandBarNoChevron: 'commandBarNoChevron_936a8d11',
    fileItem: 'fileItem_936a8d11',
    folderItem: 'folderItem_936a8d11',
    fileTypeIcon: 'fileTypeIcon_936a8d11',
    fileTypeIconIcon: 'fileTypeIconIcon_936a8d11',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_936a8d11',
    iconColumnHeader: 'iconColumnHeader_936a8d11',
    fileRow: 'fileRow_936a8d11',
    folderRow: 'folderRow_936a8d11',
    emptyFolder: 'emptyFolder_936a8d11',
    emptyFolderImage: 'emptyFolderImage_936a8d11',
    emptyFolderImageTag: 'emptyFolderImageTag_936a8d11',
    emptyFolderTitle: 'emptyFolderTitle_936a8d11',
    emptyFolderSubText: 'emptyFolderSubText_936a8d11',
    emptyFolderMobile: 'emptyFolderMobile_936a8d11',
    emptyFolderPc: 'emptyFolderPc_936a8d11',
    emptyFolderIcon: 'emptyFolderIcon_936a8d11'
};
export default styles;
/* tslint:enable */
