// @microsoft/sp-loader/lib/DeveloperTools/DeveloperTools.js
/**
 * @file DeveloperTools.ts
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 */
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import DeveloperToolsTab from './Models/DeveloperToolsTab';
import DeveloperToolsConsole from './Components/DeveloperToolsConsole';
import DeveloperToolsConsoleStore from './Stores/DeveloperToolsConsoleStore';
import { _SPFlight } from '@microsoft/sp-core-library';
import * as flightInspectorStrings from './Components/DeveloperModules/FlightInspector/FlightInspectorTab.resx';
import { FlightInspectorTab } from './Components/DeveloperModules/FlightInspector/FlightInspectorTab';
var DeveloperTools = /** @class */ (function () {
    function DeveloperTools() {
    }
    Object.defineProperty(DeveloperTools, "instance", {
        get: function () {
            if (!this._instance) {
                this._instance = new DeveloperTools();
            }
            return this._instance;
        },
        enumerable: false,
        configurable: true
    });
    DeveloperTools.prototype.initialize = function (serviceScope) {
        var developerToolsDiv = document.createElement('div');
        document.body.appendChild(developerToolsDiv);
        ReactDOM.render(React.createElement(DeveloperToolsConsole), developerToolsDiv);
        if (_SPFlight.isEnabled(466 /* SPFXDebugExperiments */)) {
            // Add flight inspector tab for debug allowed environments so it won't expose to 3rd party developers who don't need this.
            this.registerDeveloperToolsTab({
                title: flightInspectorStrings.tabTitle,
                iconName: 'Airplane',
                component: FlightInspectorTab,
                reactProps: {
                    serviceScope: serviceScope
                }
            });
        }
    };
    DeveloperTools.prototype.toggleDeveloperTools = function () {
        DeveloperToolsConsoleStore.showHideConsole(!DeveloperToolsConsoleStore.visible);
    };
    DeveloperTools.prototype.showHideDeveloperTools = function (show) {
        DeveloperToolsConsoleStore.showHideConsole(show);
    };
    DeveloperTools.prototype.registerDeveloperToolsTab = function (developerToolsTab) {
        var tab = new DeveloperToolsTab(developerToolsTab);
        DeveloperToolsConsoleStore.registerTab(tab);
    };
    return DeveloperTools;
}());
export default DeveloperTools;
