// @pnp/spfx-property-controls/lib/propertyFields/colorPicker/PropertyFieldColorPicker.js
import { PropertyPaneFieldType } from '@microsoft/sp-webpart-base';
import { getColorFromString } from 'office-ui-fabric-react/lib/utilities/color';
import * as React from 'react';
import * as ReactDom from 'react-dom';
import { setPropertyValue } from '../../helpers/GeneralHelper';
import { debounce } from '../../common/util/Util';
import { PropertyFieldColorPickerStyle, } from './IPropertyFieldColorPicker';
import PropertyFieldColorPickerHost from './PropertyFieldColorPickerHost';
var PropertyFieldColorPickerBuilder = /** @class */ (function () {
    function PropertyFieldColorPickerBuilder(_targetProperty, _properties) {
        //Properties defined by IPropertyPaneField
        this.type = PropertyPaneFieldType.Custom;
        this._debounce = debounce();
        this.targetProperty = _targetProperty;
        this.properties = {
            key: _properties.key,
            label: _properties.label,
            onPropertyChange: _properties.onPropertyChange,
            selectedColor: _properties.selectedColor,
            disabled: _properties.disabled,
            debounce: _properties.debounce,
            isHidden: _properties.isHidden,
            alphaSliderHidden: _properties.alphaSliderHidden,
            properties: _properties.properties,
            style: _properties.style,
            iconName: _properties.iconName,
            onRender: this.onRender.bind(this)
        };
        if (typeof _properties.selectedColor === 'undefined') {
            this.color = '#ffffff';
        }
        else {
            if (typeof _properties.selectedColor === 'string') {
                this.color = _properties.selectedColor;
            }
            else {
                this.color = _properties.selectedColor.str;
            }
        }
        this.valueAsObject = _properties.valueAsObject;
    }
    PropertyFieldColorPickerBuilder.prototype.render = function () {
        if (!this.elem) {
            return;
        }
        this.onRender(this.elem);
    };
    PropertyFieldColorPickerBuilder.prototype.onRender = function (elem, ctx, changeCallback) {
        if (!this.elem) {
            this.elem = elem;
        }
        this.changeCB = changeCallback;
        var element = React.createElement(PropertyFieldColorPickerHost, {
            label: this.properties.label,
            alphaSliderHidden: this.properties.alphaSliderHidden,
            disabled: this.properties.disabled,
            debounce: this.properties.debounce,
            isHidden: this.properties.isHidden,
            selectedColor: this.color,
            style: this.properties.style || PropertyFieldColorPickerStyle.Inline,
            iconName: this.properties.iconName || 'Color',
            onColorChanged: this.onColorChanged.bind(this)
        });
        ReactDom.render(element, elem);
    };
    PropertyFieldColorPickerBuilder.prototype.onColorChanged = function (newColor) {
        var _this = this;
        if (this.properties.onPropertyChange && newColor !== null) {
            var newValue_1 = (this.valueAsObject ? getColorFromString(newColor) : newColor);
            var oldValue_1 = (this.valueAsObject ? getColorFromString(this.color) : this.color);
            this.color = newColor;
            this.properties.debounce ?
                this._debounce(function () {
                    _this.onColorChangedInternal(oldValue_1, newValue_1);
                }, this.properties.debounce) :
                this.onColorChangedInternal(oldValue_1, newValue_1);
        }
    };
    PropertyFieldColorPickerBuilder.prototype.onColorChangedInternal = function (oldValue, newValue) {
        this.properties.onPropertyChange(this.targetProperty, oldValue, newValue);
        setPropertyValue(this.properties.properties, this.targetProperty, newValue);
        if (typeof this.changeCB !== 'undefined' && this.changeCB !== null) {
            this.changeCB(this.targetProperty, newValue);
        }
    };
    return PropertyFieldColorPickerBuilder;
}());
export function PropertyFieldColorPicker(targetProperty, properties) {
    return new PropertyFieldColorPickerBuilder(targetProperty, properties);
}
