// @microsoft/sp-http/lib-commonjs/index.js
"use strict";
/**
 * Base communication layer for the SharePoint Framework
 *
 * @remarks
 * This package defines the base communication layer for
 * the SharePoint Framework.  For REST calls, it handles authentication,
 * logging, diagnostics, and batching.  It also simplifies requests by
 * adding default headers that follow the recommended best practices.
 *
 * @packagedocumentation
 */
Object.defineProperty(exports, "__esModule", { value: true });
var HttpClient_1 = require("./httpClient/HttpClient");
Object.defineProperty(exports, "HttpClient", { enumerable: true, get: function () { return HttpClient_1.default; } });
var HttpClientConfiguration_1 = require("./httpClient/HttpClientConfiguration");
Object.defineProperty(exports, "HttpClientConfiguration", { enumerable: true, get: function () { return HttpClientConfiguration_1.default; } });
var HttpClientResponse_1 = require("./httpClient/HttpClientResponse");
Object.defineProperty(exports, "HttpClientResponse", { enumerable: true, get: function () { return HttpClientResponse_1.default; } });
var DigestCache_1 = require("./spHttpClient/DigestCache");
Object.defineProperty(exports, "DigestCache", { enumerable: true, get: function () { return DigestCache_1.default; } });
var SPHttpClient_1 = require("./spHttpClient/SPHttpClient");
Object.defineProperty(exports, "SPHttpClient", { enumerable: true, get: function () { return SPHttpClient_1.default; } });
var SPHttpClientCommonConfiguration_1 = require("./spHttpClient/SPHttpClientCommonConfiguration");
Object.defineProperty(exports, "SPHttpClientCommonConfiguration", { enumerable: true, get: function () { return SPHttpClientCommonConfiguration_1.default; } });
var SPHttpClientConfiguration_1 = require("./spHttpClient/SPHttpClientConfiguration");
Object.defineProperty(exports, "SPHttpClientConfiguration", { enumerable: true, get: function () { return SPHttpClientConfiguration_1.default; } });
var SPHttpClientResponse_1 = require("./spHttpClient/SPHttpClientResponse");
Object.defineProperty(exports, "SPHttpClientResponse", { enumerable: true, get: function () { return SPHttpClientResponse_1.default; } });
var SPHttpClientBatchConfiguration_1 = require("./spHttpClient/SPHttpClientBatchConfiguration");
Object.defineProperty(exports, "SPHttpClientBatchConfiguration", { enumerable: true, get: function () { return SPHttpClientBatchConfiguration_1.default; } });
var SPHttpClientBatch_1 = require("./spHttpClient/SPHttpClientBatch");
Object.defineProperty(exports, "SPHttpClientBatch", { enumerable: true, get: function () { return SPHttpClientBatch_1.default; } });
var ODataVersion_1 = require("./spHttpClient/ODataVersion");
Object.defineProperty(exports, "ODataVersion", { enumerable: true, get: function () { return ODataVersion_1.default; } });
var OAuthTokenProvider_1 = require("./oauthTokenProvider/OAuthTokenProvider");
Object.defineProperty(exports, "OAuthTokenProvider", { enumerable: true, get: function () { return OAuthTokenProvider_1.default; } });
var AadConstants_1 = require("./oauthTokenProvider/AadConstants");
Object.defineProperty(exports, "_AadConstants", { enumerable: true, get: function () { return AadConstants_1.default; } });
var AadTokenProviders_1 = require("./oauthTokenProvider/AadTokenProviders");
Object.defineProperty(exports, "_AadTokenProviders", { enumerable: true, get: function () { return AadTokenProviders_1.default; } });
var GraphHttpClient_1 = require("./graphHttpClient/GraphHttpClient");
Object.defineProperty(exports, "_GraphHttpClient", { enumerable: true, get: function () { return GraphHttpClient_1.default; } });
var GraphHttpClientResponse_1 = require("./graphHttpClient/GraphHttpClientResponse");
Object.defineProperty(exports, "_GraphHttpClientResponse", { enumerable: true, get: function () { return GraphHttpClientResponse_1.default; } });
var GraphHttpClientConfiguration_1 = require("./graphHttpClient/GraphHttpClientConfiguration");
Object.defineProperty(exports, "_GraphHttpClientConfiguration", { enumerable: true, get: function () { return GraphHttpClientConfiguration_1.default; } });
var GraphHttpClientContext_1 = require("./graphHttpClient/GraphHttpClientContext");
Object.defineProperty(exports, "_GraphHttpClientContext", { enumerable: true, get: function () { return GraphHttpClientContext_1.default; } });
var AadTokenProvider_1 = require("./oauthTokenProvider/AadTokenProvider");
Object.defineProperty(exports, "AadTokenProvider", { enumerable: true, get: function () { return AadTokenProvider_1.default; } });
Object.defineProperty(exports, "BeforeRedirectEventArgs", { enumerable: true, get: function () { return AadTokenProvider_1.BeforeRedirectEventArgs; } });
Object.defineProperty(exports, "TokenAcquisitionEventArgs", { enumerable: true, get: function () { return AadTokenProvider_1.TokenAcquisitionEventArgs; } });
var AadHttpClient_1 = require("./aadHttpClient/AadHttpClient");
Object.defineProperty(exports, "AadHttpClient", { enumerable: true, get: function () { return AadHttpClient_1.default; } });
var AadHttpClientResponse_1 = require("./aadHttpClient/AadHttpClientResponse");
Object.defineProperty(exports, "AadHttpClientResponse", { enumerable: true, get: function () { return AadHttpClientResponse_1.default; } });
var AadHttpClientConfiguration_1 = require("./aadHttpClient/AadHttpClientConfiguration");
Object.defineProperty(exports, "AadHttpClientConfiguration", { enumerable: true, get: function () { return AadHttpClientConfiguration_1.default; } });
var AadHttpClientFactory_1 = require("./aadHttpClient/AadHttpClientFactory");
Object.defineProperty(exports, "AadHttpClientFactory", { enumerable: true, get: function () { return AadHttpClientFactory_1.default; } });
var AadTokenProviderFactory_1 = require("./oauthTokenProvider/AadTokenProviderFactory");
Object.defineProperty(exports, "AadTokenProviderFactory", { enumerable: true, get: function () { return AadTokenProviderFactory_1.default; } });
var MSGraphClient_1 = require("./msGraphClient/MSGraphClient");
Object.defineProperty(exports, "MSGraphClient", { enumerable: true, get: function () { return MSGraphClient_1.default; } });
var MSGraphClientFactory_1 = require("./msGraphClient/MSGraphClientFactory");
Object.defineProperty(exports, "MSGraphClientFactory", { enumerable: true, get: function () { return MSGraphClientFactory_1.default; } });
var CacheDataProviderService_1 = require("./caching/CacheDataProviderService");
Object.defineProperty(exports, "_CacheDataProviderService", { enumerable: true, get: function () { return CacheDataProviderService_1.CacheDataProviderService; } });
Object.defineProperty(exports, "_getCacheDataProviderServiceKey", { enumerable: true, get: function () { return CacheDataProviderService_1.getCacheDataProviderServiceKey; } });
var CacheManager_1 = require("./caching/CacheManager");
Object.defineProperty(exports, "_CacheManager", { enumerable: true, get: function () { return CacheManager_1.CacheManager; } });
var IRequestCacheOptions_1 = require("./caching/IRequestCacheOptions");
Object.defineProperty(exports, "_CacheStrategy", { enumerable: true, get: function () { return IRequestCacheOptions_1.CacheStrategy; } });
