// @microsoft/sp-application-base/lib/navigator/NavigationDataError.js
// Copyright (c) Microsoft. All rights reserved.
import { __extends } from "tslib";
import { SPError, Text, Validate } from '@microsoft/sp-core-library';
/**
 * Web part error codes.
 *
 * @internal
 */
export var NavigationErrorCode;
(function (NavigationErrorCode) {
    /**
     * Failure to fetch navigation data.
     */
    NavigationErrorCode[NavigationErrorCode["fetchData"] = 0] = "fetchData";
    /**
     * Failure to parse navigation data.
     */
    NavigationErrorCode[NavigationErrorCode["parseData"] = 1] = "parseData";
    /**
     * Unknown failure
     */
    NavigationErrorCode[NavigationErrorCode["unknown"] = 2] = "unknown";
})(NavigationErrorCode || (NavigationErrorCode = {}));
/**
 * Represents the contract of a Navigation Data Error.
 * sealed
 * @internal
 */
var NavigationDataError = /** @class */ (function (_super) {
    __extends(NavigationDataError, _super);
    function NavigationDataError(errorCode, errorMessage, innerError, isExpected, errorPayload) {
        if (isExpected === void 0) { isExpected = false; }
        var _this = _super.call(this, NavigationErrorCode[errorCode], errorMessage, errorPayload) || this;
        // Manually set the prototype, as we can no longer extend built-in classes like Error, Array, Map, etc
        // tslint:disable-next-line:max-line-length
        // [https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work](https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work)
        //
        // Note: the prototype must also be set on any classes which extend this one
        _this.__proto__ = NavigationDataError.prototype; // tslint:disable-line:no-any
        _this._errorId = errorCode;
        _this.innerError = innerError;
        _this._isExpected = isExpected;
        return _this;
    }
    /**
     * Utility function to create an error response.
     * @param errorCode - Represents the navigation error code.
     * @param innerError - Represents the inner error.
     *
     * @returns An error object.
     */
    NavigationDataError.createNavigationError = function (errorCode, innerError) {
        Validate.isNotNullOrUndefined(innerError, 'innerError');
        var isExpected = NavigationDataError._isExpectedError(errorCode);
        var errorTemplate = NavigationDataError._errorTemplate(errorCode);
        var errorMessage = Text.format(errorTemplate, innerError.message);
        return new NavigationDataError(errorCode, errorMessage, innerError, isExpected);
    };
    /**
     * Utility function to create session ended reason
     * @param errorCode - Represents the navigation error code.
     * @param workload - Represents in which scenario/app the error originated
     *
     * @returns A string describing why the session is ending.
     */
    NavigationDataError.sessionEndedReason = function (errorCode, workload) {
        if (workload === void 0) { workload = 'XApp'; }
        var reason = 'RouteNav.';
        if (errorCode === NavigationErrorCode[NavigationErrorCode.parseData]) {
            reason += "Known." + workload + ".MalformedJSON";
        }
        else if (errorCode === NavigationErrorCode[NavigationErrorCode.fetchData]) {
            reason += "Known." + workload + ".NetworkError";
        }
        else {
            reason += "Unknown." + workload + ".NavigationDataError";
        }
        return reason;
    };
    NavigationDataError._isExpectedError = function (errorCode) {
        return errorCode === NavigationErrorCode.parseData;
    };
    NavigationDataError._errorTemplate = function (errorCode) {
        return NavigationDataError._errorCodeToTemplateMap[errorCode] || 'Failed to navigate. Error: {0}';
    };
    Object.defineProperty(NavigationDataError.prototype, "id", {
        /**
         * Gets the error Id.
         */
        get: function () {
            return this._errorId;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NavigationDataError.prototype, "category", {
        /**
         * Gets the error category.
         */
        get: function () {
            return 'NavigationDataError';
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NavigationDataError.prototype, "isExpected", {
        /**
         * Gets a value indicating whether the error is expected.
         */
        get: function () {
            return this._isExpected;
        },
        enumerable: false,
        configurable: true
    });
    NavigationDataError._errorCodeToTemplateMap = new Map([
        [NavigationErrorCode.fetchData, 'Failed to retrieve navigation data. Error: {0}'],
        [NavigationErrorCode.parseData, 'Failed to parse navigation data. Error: {0}']
    ]);
    return NavigationDataError;
}(SPError));
export default NavigationDataError;
