// @microsoft/office-ui-fabric-react-bundle/lib-commonjs/office-ui-fabric-react-bundle.js
"use strict";
/* tslint:disable:export-name variable-name */
Object.defineProperty(exports, "__esModule", { value: true });
exports.workaround_PopupIndex = exports.workaround_PopupComponent = exports.workaround_Popup = exports.workaround_CheckIndex = exports.workaround_CheckComponent = exports.workaround_Check = exports.workaround_ImageComponent = exports.workaround_IconIndex = exports.workaround_IconComponent = exports.workaround_Icon = exports.workaround_withViewport = exports.workaround_withResponsiveMode = exports.initializeIcons = void 0;
var tslib_1 = require("tslib");
/**
 * Initializing the OUFR icons.
 *
 * This should happen only once and before any component starts using icons.
 */
var icons_1 = require("@uifabric/icons");
// Register icons and pull the fonts from the default SharePoint cdn.
icons_1.initializeIcons(); // 47KB minified
/**
 * Export a fake "initializeIcons" function to satisfy external dependencies
 * that call initializeIcons();
 *
 * @public
 */
function initializeIcons() {
    // no-op
}
exports.initializeIcons = initializeIcons;
/**
 * office-ui-fabric-react control exports.
 *
 * We are exposing only those controls, which are used in the sp-pages assembly.
 */
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Button"), exports); // 1KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Callout"), exports); // 1KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/CommandBar"), exports); // 18KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/ContextualMenu"), exports); // 1KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Fabric"), exports); // 1KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/FocusTrapZone"), exports); // 1KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/FocusZone"), exports); // 1KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Icon"), exports); // 1KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Image"), exports); // 1KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Label"), exports); // 1KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Link"), exports); // 3KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/List"), exports); // 9KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/MessageBar"), exports); // 13KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Overlay"), exports); // 1KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Persona"), exports); // 9KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Spinner"), exports); // 5KB minified
// Explicit exports to avoid unnecessarily re-exporting large MaskedTextField
tslib_1.__exportStar(require("office-ui-fabric-react/lib/components/TextField/TextField"), exports);
tslib_1.__exportStar(require("office-ui-fabric-react/lib/components/TextField/TextField.base"), exports);
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Utilities"), exports); // 1KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/components/Tooltip/TooltipHost"), exports); // 1KB minified
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Shimmer"), exports);
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Foundation"), exports);
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Styling"), exports);
tslib_1.__exportStar(require("office-ui-fabric-react/lib/Stack"), exports);
// 24KB minified
// Temporary workarounds to eliminate extra bundling and webpack warnings
// Fix the underlying issues: VSO#274960
var withResponsiveMode = tslib_1.__importStar(require("office-ui-fabric-react/lib/utilities/decorators/withResponsiveMode"));
/**
 * @internal
 */
exports.workaround_withResponsiveMode = withResponsiveMode;
var withViewport = tslib_1.__importStar(require("office-ui-fabric-react/lib/utilities/decorators/withViewport"));
/**
 * @internal
 */
exports.workaround_withViewport = withViewport;
// Icon workarounds
var Icon = tslib_1.__importStar(require("office-ui-fabric-react/lib/Icon"));
/**
 * @internal
 */
exports.workaround_Icon = Icon;
var IconComponent = tslib_1.__importStar(require("office-ui-fabric-react/lib/components/Icon"));
/**
 * @internal
 */
exports.workaround_IconComponent = IconComponent;
var IconIndex = tslib_1.__importStar(require("office-ui-fabric-react/lib/components/Icon/index"));
/**
 * @internal
 */
exports.workaround_IconIndex = IconIndex;
var ImageComponent = tslib_1.__importStar(require("office-ui-fabric-react/lib/components/Image/Image"));
/**
 * @internal
 */
exports.workaround_ImageComponent = ImageComponent;
// Check workarounds
var Check = tslib_1.__importStar(require("office-ui-fabric-react/lib/Check"));
/**
 * @internal
 */
exports.workaround_Check = Check;
var CheckComponent = tslib_1.__importStar(require("office-ui-fabric-react/lib/components/Check/Check"));
/**
 * @internal
 */
exports.workaround_CheckComponent = CheckComponent;
var CheckIndex = tslib_1.__importStar(require("office-ui-fabric-react/lib/components/Check/index"));
/**
 * @internal
 */
exports.workaround_CheckIndex = CheckIndex;
// Popup workarounds
var Popup = tslib_1.__importStar(require("office-ui-fabric-react/lib/Popup"));
/**
 * @internal
 */
exports.workaround_Popup = Popup;
var PopupComponent = tslib_1.__importStar(require("office-ui-fabric-react/lib/components/Popup/Popup"));
/**
 * @internal
 */
exports.workaround_PopupComponent = PopupComponent;
var PopupIndex = tslib_1.__importStar(require("office-ui-fabric-react/lib/components/Popup/index"));
/**
 * @internal
 */
exports.workaround_PopupIndex = PopupIndex;
tslib_1.__exportStar(require("./autobind"), exports);
