// @microsoft/sp-build-core-tasks/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.serve = exports.configureRig = exports.deployAzureStorage = exports.packageSolution = exports.configureWebpack = void 0;
const ConfigureWebpackTask_1 = require("./webpack/ConfigureWebpackTask");
const PackageSolutionTask_1 = require("./packageSolution/PackageSolutionTask");
const DeployAzureStorageTask_1 = require("./deployAzureStorage/DeployAzureStorageTask");
const ConfigureRigTask_1 = require("./configJson/ConfigureRigTask");
const SpfxServeTask_1 = require("./spfxServe/SpfxServeTask");
var GulpTaskTerminalProvider_1 = require("./utilities/GulpTaskTerminalProvider");
Object.defineProperty(exports, "GulpTaskTerminalProvider", { enumerable: true, get: function () { return GulpTaskTerminalProvider_1.GulpTaskTerminalProvider; } });
__exportStar(require("./cumulativeManifestProcessor/GulpTaskCumulativeManifestProcessor"), exports);
exports.configureWebpack = new ConfigureWebpackTask_1.ConfigureWebpackTask();
exports.packageSolution = new PackageSolutionTask_1.PackageSolutionTask();
exports.deployAzureStorage = new DeployAzureStorageTask_1.DeployAzureStorageTask();
exports.configureRig = new ConfigureRigTask_1.ConfigureRigTask();
exports.serve = new SpfxServeTask_1.SpfxServeTask();
