// @microsoft/sp-loader/lib/DeveloperTools/Stores/DeveloperToolsConsoleStore.js
/**
 * @file DeveloperToolsConsoleStore.ts
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 */
import { __extends } from "tslib";
import BaseStore from './BaseStore';
import ManifestManager from './../Components/DeveloperModules/ManifestManager/ManifestManager';
import PerformanceDisplay from './../Components/DeveloperModules/PerformanceDisplay/PerformanceDisplay';
import TraceDisplay from './../Components/DeveloperModules/TraceDisplay/TraceDisplay';
import DeveloperToolsTab from './../Models/DeveloperToolsTab';
import * as manifestManagerStrings from './../Components/DeveloperModules/ManifestManager/ManifestManager.resx';
import * as perfDisplayStrings from './../Components/DeveloperModules/PerformanceDisplay/PerformanceDisplay.resx';
import * as traceDisplayStrings from './../Components/DeveloperModules/TraceDisplay/TraceDisplay.resx';
var DEFAULT_HEIGHT = 1 / 4;
var MIN_HEIGHT = 1 / 6;
var MAX_HEIGHT = 3 / 4;
var DeveloperToolsConsoleStore = /** @class */ (function (_super) {
    __extends(DeveloperToolsConsoleStore, _super);
    function DeveloperToolsConsoleStore() {
        var _this = _super.call(this) || this;
        _this._height = DEFAULT_HEIGHT;
        _this._visible = false;
        _this._selectedTabId = 0;
        _this._tabs = [
            new DeveloperToolsTab({
                title: traceDisplayStrings.TabTitle,
                iconName: 'ShowResults',
                component: TraceDisplay
            }),
            new DeveloperToolsTab({
                title: perfDisplayStrings.TabTitle,
                iconName: 'SpeedHigh',
                component: PerformanceDisplay
            }),
            new DeveloperToolsTab({
                title: manifestManagerStrings.tabTitle,
                iconName: 'Script',
                component: ManifestManager
            })
        ];
        return _this;
    }
    Object.defineProperty(DeveloperToolsConsoleStore, "instance", {
        get: function () {
            if (!this._instance) {
                this._instance = new DeveloperToolsConsoleStore();
            }
            return this._instance;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DeveloperToolsConsoleStore.prototype, "height", {
        get: function () {
            return this._height;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DeveloperToolsConsoleStore.prototype, "visible", {
        get: function () {
            return this._visible;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DeveloperToolsConsoleStore.prototype, "selectedTabId", {
        get: function () {
            return this._selectedTabId;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DeveloperToolsConsoleStore.prototype, "tabs", {
        get: function () {
            return this._tabs;
        },
        enumerable: false,
        configurable: true
    });
    DeveloperToolsConsoleStore.prototype.resizeConsole = function (height) {
        var newHeight = this._height;
        if (height >= MAX_HEIGHT) {
            newHeight = MAX_HEIGHT;
        }
        else if (height <= MIN_HEIGHT) {
            newHeight = MIN_HEIGHT;
        }
        else {
            newHeight = height;
        }
        if (newHeight !== this._height) {
            this._height = newHeight;
            this.emitChange();
        }
    };
    DeveloperToolsConsoleStore.prototype.showHideConsole = function (show) {
        if (this._visible !== show) {
            this._visible = show;
            this.emitChange();
        }
    };
    DeveloperToolsConsoleStore.prototype.registerTab = function (tab) {
        this._tabs.push(tab);
        if (this._tabs.length === 1) {
            this._selectedTabId = 0;
        }
        this.emitChange();
    };
    return DeveloperToolsConsoleStore;
}(BaseStore));
export { DeveloperToolsConsoleStore };
// instantiate the store so it can start listening to dispatch events
export default DeveloperToolsConsoleStore.instance; // tslint:disable-line:export-name
