// @microsoft/sp-application-base/lib/extensibility/SearchQueryManager.js
import { ServiceKey, _SPFlight } from '@microsoft/sp-core-library';
import { _QosMonitor } from '@microsoft/sp-diagnostics';
import { SPComponentLoader } from '@microsoft/sp-loader';
import CustomActionLocations from './customAction/CustomActionLocations';
import { shellSearch } from '@ms/sp-suite-nav';
/**
 * Search query manager is used to get the search query in SPFx.
 * It supports Search query modifier extensions that can enrich the search query, and supports cases where the user
 * comes from classic pages to modern.
 *
 * @internal
 */
var SearchQueryManager = /** @class */ (function () {
    function SearchQueryManager(serviceScope) {
        this._isInitialized = false;
        this._areExtensionsLoaded = false;
        this._serviceScope = serviceScope;
    }
    SearchQueryManager.prototype.initialize = function (preloadedData) {
        var qosMonitor = new _QosMonitor('SearchQueryManager.initialize');
        this._preloadedData = preloadedData;
        var queryParameters = new URL(window.location.href).searchParams;
        this._isInitialized = false;
        this._areExtensionsLoaded = false;
        var searchQuery = this._getUrlSearchQuery(queryParameters);
        if (searchQuery) {
            if (!_SPFlight.isEnabled(1136 /* O365ShellModuleJSClient */)) {
                // The faster suiteNav sets the searchBox text from the server-side, so this is unnecessary.
                this._updateSearchBoxText(searchQuery);
            }
        }
        if (searchQuery && this._hasExtensions(queryParameters)) {
            return this._initializeExtensions(qosMonitor);
        }
        else {
            this._isInitialized = true;
            this._areExtensionsLoaded = false;
            qosMonitor.writeSuccess({ numberOfExtensions: 0 });
            return Promise.resolve();
        }
    };
    /**
     * Update SearchQueryManager from location
     * @param location - next location Data
     */
    SearchQueryManager.prototype.updateFromLocation = function (location) {
        var qosMonitor = new _QosMonitor('SearchQueryManager.update');
        var queryParameters = new URLSearchParams(location.search.substring(1));
        var searchQuery = this._getUrlSearchQuery(queryParameters);
        if (searchQuery) {
            this._updateSearchBoxText(searchQuery);
        }
        if (searchQuery && this._hasExtensions(queryParameters) && !this._areExtensionsLoaded) {
            return this._initializeExtensions(qosMonitor);
        }
        else {
            qosMonitor.writeSuccess({ numberOfExtensions: 0 });
            return Promise.resolve();
        }
    };
    SearchQueryManager.prototype.isInitialized = function () {
        return this._isInitialized;
    };
    /**
     * Get the search query from the current URL.
     */
    SearchQueryManager.prototype.getSearchQuery = function () {
        if (!this._isInitialized) {
            return Promise.reject(new Error('SearchQueryManager is not initialized'));
        }
        var qosMonitor = new _QosMonitor('SearchQueryManager.getSearchQuery');
        var queryParameters = new URL(window.location.href).searchParams;
        var searchQuery = this._getUrlSearchQuery(queryParameters);
        if (searchQuery && this._hasExtensions(queryParameters)) {
            return this._getExtensionManager()
                .then(function (extensionManager) { return extensionManager.getSearchQuery(searchQuery); })
                .then(function (q) {
                qosMonitor.writeSuccess();
                return q;
            })
                .catch(function (e) {
                qosMonitor.writeUnexpectedFailure(undefined, e);
                throw e;
            });
        }
        else {
            qosMonitor.writeSuccess();
            return Promise.resolve(searchQuery);
        }
    };
    SearchQueryManager.prototype._getUrlSearchQuery = function (searchParams) {
        // If ?q is not present, ?k can be used (support cases coming from classic pages)
        return searchParams.get('q') || searchParams.get('k') || undefined;
    };
    /**
     * Returns true if there is any SearchQueryModifier extension.
     * When using query params, it just checks if there is any custom action regardless of type.
     * This is exclusively a debug scenario and it's faster and smaller code, considering this runs in a core scenario.
     */
    SearchQueryManager.prototype._hasExtensions = function (searchParams) {
        return (searchParams.has('customActions') ||
            (!!this._preloadedData.customActions &&
                this._preloadedData.customActions.some(function (customAction) { return customAction.location === CustomActionLocations.SEARCH_QUERY_MODIFIER; })));
    };
    SearchQueryManager.prototype._getExtensionManager = function () {
        var _this = this;
        if (!this._extensionManagerPromise) {
            this._extensionManagerPromise = SPComponentLoader.loadComponentById('4958ea79-6ff3-4480-8291-0932dd010869').then(function (spSearchExtensibility) {
                return new spSearchExtensibility._SearchQueryExtensionManager(_this._serviceScope);
            });
        }
        return this._extensionManagerPromise;
    };
    SearchQueryManager.prototype._initializeExtensions = function (qosMonitor) {
        var _this = this;
        return this._getExtensionManager()
            .then(function (extensionManager) {
            return extensionManager
                .initializeExtensions(_this._preloadedData.customActions)
                .then(function (numberOfExtensions) {
                _this._areExtensionsLoaded = true;
                _this._isInitialized = true;
                qosMonitor.writeSuccess({ numberOfExtensions: numberOfExtensions });
            });
        })
            .catch(function (e) {
            qosMonitor.writeUnexpectedFailure(undefined, e, { numberOfExtensions: 0 });
            throw e;
        });
    };
    /**
     * Updates the search box text.
     * If there is no search query it doesn't do anything to preserve what the user originally typed.
     */
    SearchQueryManager.prototype._updateSearchBoxText = function (searchQuery) {
        // VSO:887348 - Check that current searchBox text is the same as the previous page query param before updating.
        if (!!searchQuery && this._serviceScope) {
            shellSearch(this._serviceScope).then(function (search) {
                search.SetSearchText(searchQuery);
            });
        }
    };
    SearchQueryManager.serviceKey = ServiceKey.create('sp-application-base:SearchQueryManager', SearchQueryManager);
    return SearchQueryManager;
}());
export default SearchQueryManager;
