// @microsoft/office-ui-fabric-react-bundle/lib/office-ui-fabric-react-bundle.js
/* tslint:disable:export-name variable-name */
/**
 * Initializing the OUFR icons.
 *
 * This should happen only once and before any component starts using icons.
 */
import { initializeIcons as realInitializeIcons } from '@uifabric/icons';
// Register icons and pull the fonts from the default SharePoint cdn.
realInitializeIcons(); // 47KB minified
/**
 * Export a fake "initializeIcons" function to satisfy external dependencies
 * that call initializeIcons();
 *
 * @public
 */
export function initializeIcons() {
    // no-op
}
/**
 * office-ui-fabric-react control exports.
 *
 * We are exposing only those controls, which are used in the sp-pages assembly.
 */
export * from 'office-ui-fabric-react/lib/Button'; // 1KB minified
export * from 'office-ui-fabric-react/lib/Callout'; // 1KB minified
export * from 'office-ui-fabric-react/lib/CommandBar'; // 18KB minified
export * from 'office-ui-fabric-react/lib/ContextualMenu'; // 1KB minified
export * from 'office-ui-fabric-react/lib/Fabric'; // 1KB minified
export * from 'office-ui-fabric-react/lib/FocusTrapZone'; // 1KB minified
export * from 'office-ui-fabric-react/lib/FocusZone'; // 1KB minified
export * from 'office-ui-fabric-react/lib/Icon'; // 1KB minified
export * from 'office-ui-fabric-react/lib/Image'; // 1KB minified
export * from 'office-ui-fabric-react/lib/Label'; // 1KB minified
export * from 'office-ui-fabric-react/lib/Link'; // 3KB minified
export * from 'office-ui-fabric-react/lib/List'; // 9KB minified
export * from 'office-ui-fabric-react/lib/MessageBar'; // 13KB minified
export * from 'office-ui-fabric-react/lib/Overlay'; // 1KB minified
export * from 'office-ui-fabric-react/lib/Persona'; // 9KB minified
export * from 'office-ui-fabric-react/lib/Spinner'; // 5KB minified
// Explicit exports to avoid unnecessarily re-exporting large MaskedTextField
export * from 'office-ui-fabric-react/lib/components/TextField/TextField';
export * from 'office-ui-fabric-react/lib/components/TextField/TextField.base';
export * from 'office-ui-fabric-react/lib/Utilities'; // 1KB minified
export * from 'office-ui-fabric-react/lib/components/Tooltip/TooltipHost'; // 1KB minified
export * from 'office-ui-fabric-react/lib/Shimmer';
export * from 'office-ui-fabric-react/lib/Foundation';
export * from 'office-ui-fabric-react/lib/Styling';
export * from 'office-ui-fabric-react/lib/Stack';
// 24KB minified
// Temporary workarounds to eliminate extra bundling and webpack warnings
// Fix the underlying issues: VSO#274960
import * as withResponsiveMode from 'office-ui-fabric-react/lib/utilities/decorators/withResponsiveMode';
/**
 * @internal
 */
export var workaround_withResponsiveMode = withResponsiveMode;
import * as withViewport from 'office-ui-fabric-react/lib/utilities/decorators/withViewport';
/**
 * @internal
 */
export var workaround_withViewport = withViewport;
// Icon workarounds
import * as Icon from 'office-ui-fabric-react/lib/Icon';
/**
 * @internal
 */
export var workaround_Icon = Icon;
import * as IconComponent from 'office-ui-fabric-react/lib/components/Icon';
/**
 * @internal
 */
export var workaround_IconComponent = IconComponent;
import * as IconIndex from 'office-ui-fabric-react/lib/components/Icon/index';
/**
 * @internal
 */
export var workaround_IconIndex = IconIndex;
import * as ImageComponent from 'office-ui-fabric-react/lib/components/Image/Image';
/**
 * @internal
 */
export var workaround_ImageComponent = ImageComponent;
// Check workarounds
import * as Check from 'office-ui-fabric-react/lib/Check';
/**
 * @internal
 */
export var workaround_Check = Check;
import * as CheckComponent from 'office-ui-fabric-react/lib/components/Check/Check';
/**
 * @internal
 */
export var workaround_CheckComponent = CheckComponent;
import * as CheckIndex from 'office-ui-fabric-react/lib/components/Check/index';
/**
 * @internal
 */
export var workaround_CheckIndex = CheckIndex;
// Popup workarounds
import * as Popup from 'office-ui-fabric-react/lib/Popup';
/**
 * @internal
 */
export var workaround_Popup = Popup;
import * as PopupComponent from 'office-ui-fabric-react/lib/components/Popup/Popup';
/**
 * @internal
 */
export var workaround_PopupComponent = PopupComponent;
import * as PopupIndex from 'office-ui-fabric-react/lib/components/Popup/index';
/**
 * @internal
 */
export var workaround_PopupIndex = PopupIndex;
export * from './autobind';
