// @microsoft/spfx-heft-plugins/lib/plugins/webpackConfigurationPlugin/WebpackConfigurationPlugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebpackConfigurationPlugin = void 0;
const path = __importStar(require("path"));
const node_core_library_1 = require("@rushstack/node-core-library");
const WebpackConfigurationGenerator_1 = require("./WebpackConfigurationGenerator");
const configJsonReader_1 = require("../../spfxConfig/configJsonReader");
const configJsonMigrator_1 = require("../../spfxConfig/configJsonMigrator");
class WebpackConfigurationPlugin {
    constructor() {
        this.pluginName = "spfx-webpack-configuration-plugin" /* SpfxWebpackConfigurationPlugin */;
    }
    get accessor() {
        if (!this._accessor) {
            this._accessor = {
                properties: {}
            };
        }
        return this._accessor;
    }
    /**
     * Generate the SPFx component webpack configuration
     */
    apply(heftSession, heftConfiguration) {
        heftSession.hooks.build.tap("spfx-webpack-configuration-plugin" /* SpfxWebpackConfigurationPlugin */, (build) => {
            build.hooks.bundle.tap("spfx-webpack-configuration-plugin" /* SpfxWebpackConfigurationPlugin */, (bundleSubstage) => {
                bundleSubstage.hooks.configureWebpack.tapPromise("spfx-webpack-configuration-plugin" /* SpfxWebpackConfigurationPlugin */, async (webpackConfiguration) => {
                    const scopedLogger = heftSession.requestScopedLogger('spfx-webpack-configuration-plugin');
                    if (webpackConfiguration) {
                        scopedLogger.emitError(new Error('A webpack configuration must not already have been provided.'));
                    }
                    return await this._generateSpfxWebpackConfigurationAsync(heftConfiguration, build, scopedLogger);
                });
            });
        });
    }
    async _generateSpfxWebpackConfigurationAsync(heftConfiguration, build, scopedLogger) {
        const configJson = this._loadConfigJson(heftConfiguration, scopedLogger);
        if (configJson) {
            const { properties: accessorProperties } = this.accessor;
            return (await WebpackConfigurationGenerator_1.WebpackConfigurationGenerator.generateWebpackConfigurationAsync({
                configJson: configJson,
                folders: {
                    buildFolder: heftConfiguration.buildFolder,
                    tempFolder: path.join(heftConfiguration.buildFolder, 'temp'),
                    releaseFolder: path.join(heftConfiguration.buildFolder, 'release'),
                    outputFolder: path.join(heftConfiguration.buildFolder, 'dist')
                },
                terminal: scopedLogger.terminal,
                production: build.properties.production,
                serveMode: build.properties.serveMode,
                projectPackageJson: heftConfiguration.projectPackageJson,
                localization: accessorProperties.localization,
                singleLocale: build.properties.locale,
                linkedExternalsToBundle: accessorProperties.linkedExternalsToBundle,
                libraryName: accessorProperties.libraryName,
                libraryTarget: accessorProperties.libraryTarget,
                disableAsyncComponentLoading: accessorProperties.disableAsyncComponentLoading,
                assetsAreVersioned: accessorProperties.assetsAreVersioned,
                includeScssLoaders: true,
                generateCssClassName: accessorProperties.generateCssClassName,
                loadThemedStylesImportPath: accessorProperties.loadThemedStylesImportPath
            }));
        }
        else {
            return undefined;
        }
    }
    _loadConfigJson(heftConfiguration, scopedLogger) {
        const configFilePath = path.join(heftConfiguration.buildFolder, 'config', 'config.json');
        const configResult = configJsonReader_1.readConfigFile(scopedLogger.terminal.writeVerboseLine.bind(scopedLogger.terminal), configFilePath);
        if (configResult.missingFile) {
            scopedLogger.emitWarning(new Error('Missing config.json file. If this is a non-web project, consider using a non-web build rig ' +
                'like @microsoft/sp-build-node'));
        }
        if (configResult.readError) {
            scopedLogger.emitError(new Error(`Error reading config.json file: ${configResult.readError}`));
            return undefined;
        }
        if (!configResult.isLatestVersion) {
            configJsonMigrator_1.migrateToLatestVersion(scopedLogger.terminal.writeWarningLine.bind(scopedLogger.terminal), configResult.configData);
            node_core_library_1.JsonFile.save(configResult.configData, configFilePath, { prettyFormatting: true });
            scopedLogger.emitWarning(new Error(`This project is using an old config.json file version (${configResult.version}). The ` +
                'config.json file has been migrated to the latest version.'));
        }
        return configResult.configData;
    }
}
exports.WebpackConfigurationPlugin = WebpackConfigurationPlugin;
exports.default = new WebpackConfigurationPlugin();
