// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/RecentFilesTab/RecentFilesTab.module.scss.js
/* tslint:disable */
require("./RecentFilesTab.module.css");
var styles = {
    filePicker: 'filePicker_78e3e848',
    tabsContainer: 'tabsContainer_78e3e848',
    tabContainer: 'tabContainer_78e3e848',
    nav: 'nav_78e3e848',
    tabHeaderContainer: 'tabHeaderContainer_78e3e848',
    breadcrumbNav: 'breadcrumbNav_78e3e848',
    breadcrumbNavItem: 'breadcrumbNavItem_78e3e848',
    tab: 'tab_78e3e848',
    tabOffset: 'tabOffset_78e3e848',
    tabFiles: 'tabFiles_78e3e848',
    scrollablePaneWrapper: 'scrollablePaneWrapper_78e3e848',
    actionButtonsContainer: 'actionButtonsContainer_78e3e848',
    actionButtons: 'actionButtons_78e3e848',
    actionButton: 'actionButton_78e3e848',
    tabHeader: 'tabHeader_78e3e848',
    selectedTab: 'selectedTab_78e3e848',
    header: 'header_78e3e848',
    xlg: 'xlg_78e3e848',
    xxlg: 'xxlg_78e3e848',
    xxxlg: 'xxxlg_78e3e848',
    xxxxlg: 'xxxxlg_78e3e848',
    itemPickerTopBar: 'itemPickerTopBar_78e3e848',
    commandBarNoChevron: 'commandBarNoChevron_78e3e848',
    fileItem: 'fileItem_78e3e848',
    folderItem: 'folderItem_78e3e848',
    fileTypeIcon: 'fileTypeIcon_78e3e848',
    fileTypeIconIcon: 'fileTypeIconIcon_78e3e848',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_78e3e848',
    iconColumnHeader: 'iconColumnHeader_78e3e848',
    fileRow: 'fileRow_78e3e848',
    folderRow: 'folderRow_78e3e848',
    emptyFolder: 'emptyFolder_78e3e848',
    emptyFolderImage: 'emptyFolderImage_78e3e848',
    emptyFolderImageTag: 'emptyFolderImageTag_78e3e848',
    emptyFolderTitle: 'emptyFolderTitle_78e3e848',
    emptyFolderSubText: 'emptyFolderSubText_78e3e848',
    emptyFolderMobile: 'emptyFolderMobile_78e3e848',
    emptyFolderPc: 'emptyFolderPc_78e3e848',
    emptyFolderIcon: 'emptyFolderIcon_78e3e848',
    gridListCell: 'gridListCell_78e3e848',
    cellRight: 'cellRight_78e3e848',
    cellLastRow: 'cellLastRow_78e3e848',
    itemTile: 'itemTile_78e3e848',
    isFile: 'isFile_78e3e848',
    isPhoto: 'isPhoto_78e3e848',
    isVideo: 'isVideo_78e3e848',
    hasThumbnail: 'hasThumbnail_78e3e848',
    itemTileContent: 'itemTileContent_78e3e848',
    itemTileDiscoverFileContainer: 'itemTileDiscoverFileContainer_78e3e848',
    itemTileFileContainer: 'itemTileFileContainer_78e3e848',
    itemTileFile: 'itemTileFile_78e3e848',
    itemTileThumbnail: 'itemTileThumbnail_78e3e848',
    image: 'image_78e3e848',
    itemTileCheckCircle: 'itemTileCheckCircle_78e3e848',
    isSelected: 'isSelected_78e3e848',
    itemTileNamePlate: 'itemTileNamePlate_78e3e848',
    itemTileName: 'itemTileName_78e3e848',
    itemTileChildCount: 'itemTileChildCount_78e3e848',
    itemTileSubText: 'itemTileSubText_78e3e848',
    recentGridList: 'recentGridList_78e3e848'
};
export default styles;
/* tslint:enable */
