// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/controls/DocumentTile/DocumentTile.module.scss.js
/* tslint:disable */
require("./DocumentTile.module.css");
var styles = {
    filePicker: 'filePicker_0951efe5',
    tabsContainer: 'tabsContainer_0951efe5',
    tabContainer: 'tabContainer_0951efe5',
    nav: 'nav_0951efe5',
    tabHeaderContainer: 'tabHeaderContainer_0951efe5',
    breadcrumbNav: 'breadcrumbNav_0951efe5',
    breadcrumbNavItem: 'breadcrumbNavItem_0951efe5',
    tab: 'tab_0951efe5',
    tabOffset: 'tabOffset_0951efe5',
    tabFiles: 'tabFiles_0951efe5',
    scrollablePaneWrapper: 'scrollablePaneWrapper_0951efe5',
    actionButtonsContainer: 'actionButtonsContainer_0951efe5',
    actionButtons: 'actionButtons_0951efe5',
    actionButton: 'actionButton_0951efe5',
    tabHeader: 'tabHeader_0951efe5',
    selectedTab: 'selectedTab_0951efe5',
    header: 'header_0951efe5',
    xlg: 'xlg_0951efe5',
    xxlg: 'xxlg_0951efe5',
    xxxlg: 'xxxlg_0951efe5',
    xxxxlg: 'xxxxlg_0951efe5',
    itemPickerTopBar: 'itemPickerTopBar_0951efe5',
    commandBarNoChevron: 'commandBarNoChevron_0951efe5',
    fileItem: 'fileItem_0951efe5',
    folderItem: 'folderItem_0951efe5',
    fileTypeIcon: 'fileTypeIcon_0951efe5',
    fileTypeIconIcon: 'fileTypeIconIcon_0951efe5',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_0951efe5',
    iconColumnHeader: 'iconColumnHeader_0951efe5',
    fileRow: 'fileRow_0951efe5',
    folderRow: 'folderRow_0951efe5',
    emptyFolder: 'emptyFolder_0951efe5',
    emptyFolderImage: 'emptyFolderImage_0951efe5',
    emptyFolderImageTag: 'emptyFolderImageTag_0951efe5',
    emptyFolderTitle: 'emptyFolderTitle_0951efe5',
    emptyFolderSubText: 'emptyFolderSubText_0951efe5',
    emptyFolderMobile: 'emptyFolderMobile_0951efe5',
    emptyFolderPc: 'emptyFolderPc_0951efe5',
    emptyFolderIcon: 'emptyFolderIcon_0951efe5',
    aboveNameplate: 'aboveNameplate_0951efe5',
    tileContent: 'tileContent_0951efe5',
    foreground: 'foreground_0951efe5',
    grid: 'grid_0951efe5',
    cellContent: 'cellContent_0951efe5',
    content: 'content_0951efe5',
    tile: 'tile_0951efe5',
    invokable: 'invokable_0951efe5',
    namePlate: 'namePlate_0951efe5',
    nameplate: 'nameplate_0951efe5',
    name: 'name_0951efe5',
    isLarge: 'isLarge_0951efe5',
    activity: 'activity_0951efe5',
    signalField: 'signalField_0951efe5',
    compact: 'compact_0951efe5',
    signalFieldValue: 'signalFieldValue_0951efe5',
    folderList: 'folderList_0951efe5',
    link: 'link_0951efe5',
    cell: 'cell_0951efe5',
    listCell: 'listCell_0951efe5',
    label: 'label_0951efe5',
    check: 'check_0951efe5',
    selectable: 'selectable_0951efe5',
    showCheck: 'showCheck_0951efe5',
    selected: 'selected_0951efe5',
    dialogMainOverride: 'dialogMainOverride_0951efe5',
    itemTileBlankCover: 'itemTileBlankCover_0951efe5',
    metadata: 'metadata_0951efe5',
    isSmall: 'isSmall_0951efe5',
    listPage: 'listPage_0951efe5',
    odItemTile2FolderCover: 'odItemTile2FolderCover_0951efe5',
    standaloneListBreadcrumb: 'standaloneListBreadcrumb_0951efe5',
    odItemTile2Image: 'odItemTile2Image_0951efe5',
    odImageFrame2: 'odImageFrame2_0951efe5',
    odImageFrame2Image: 'odImageFrame2Image_0951efe5',
    odImageFrame: 'odImageFrame_0951efe5',
    odImageStack: 'odImageStack_0951efe5',
    odImageStackTile: 'odImageStackTile_0951efe5',
    odImageTile: 'odImageTile_0951efe5',
    odImageTileBackground: 'odImageTileBackground_0951efe5',
    odImageTileImage: 'odImageTileImage_0951efe5',
    odItemTile2SmallIcon: 'odItemTile2SmallIcon_0951efe5'
};
export default styles;
/* tslint:enable */
