// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/controls/FileBrowser/FileBrowser.module.scss.js
/* tslint:disable */
require("./FileBrowser.module.css");
var styles = {
    filePicker: 'filePicker_33a5f558',
    tabsContainer: 'tabsContainer_33a5f558',
    tabContainer: 'tabContainer_33a5f558',
    nav: 'nav_33a5f558',
    tabHeaderContainer: 'tabHeaderContainer_33a5f558',
    breadcrumbNav: 'breadcrumbNav_33a5f558',
    breadcrumbNavItem: 'breadcrumbNavItem_33a5f558',
    tab: 'tab_33a5f558',
    tabOffset: 'tabOffset_33a5f558',
    tabFiles: 'tabFiles_33a5f558',
    scrollablePaneWrapper: 'scrollablePaneWrapper_33a5f558',
    actionButtonsContainer: 'actionButtonsContainer_33a5f558',
    actionButtons: 'actionButtons_33a5f558',
    actionButton: 'actionButton_33a5f558',
    tabHeader: 'tabHeader_33a5f558',
    selectedTab: 'selectedTab_33a5f558',
    header: 'header_33a5f558',
    xlg: 'xlg_33a5f558',
    xxlg: 'xxlg_33a5f558',
    xxxlg: 'xxxlg_33a5f558',
    xxxxlg: 'xxxxlg_33a5f558',
    itemPickerTopBar: 'itemPickerTopBar_33a5f558',
    commandBarNoChevron: 'commandBarNoChevron_33a5f558',
    fileItem: 'fileItem_33a5f558',
    folderItem: 'folderItem_33a5f558',
    fileTypeIcon: 'fileTypeIcon_33a5f558',
    fileTypeIconIcon: 'fileTypeIconIcon_33a5f558',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_33a5f558',
    iconColumnHeader: 'iconColumnHeader_33a5f558',
    fileRow: 'fileRow_33a5f558',
    folderRow: 'folderRow_33a5f558',
    emptyFolder: 'emptyFolder_33a5f558',
    emptyFolderImage: 'emptyFolderImage_33a5f558',
    emptyFolderImageTag: 'emptyFolderImageTag_33a5f558',
    emptyFolderTitle: 'emptyFolderTitle_33a5f558',
    emptyFolderSubText: 'emptyFolderSubText_33a5f558',
    emptyFolderMobile: 'emptyFolderMobile_33a5f558',
    emptyFolderPc: 'emptyFolderPc_33a5f558',
    emptyFolderIcon: 'emptyFolderIcon_33a5f558'
};
export default styles;
/* tslint:enable */
