// @microsoft/spfx-heft-plugins/lib/spfxManifests/manifestFileBuilder/ManifestsFileBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManifestsFileBuilder = void 0;
const webpack_1 = __importDefault(require("webpack"));
const path = __importStar(require("path"));
const lodash_1 = require("lodash");
const set_webpack_public_path_plugin_1 = require("@rushstack/set-webpack-public-path-plugin");
const node_core_library_1 = require("@rushstack/node-core-library");
/**
 * @internal
 */
class ManifestsFileBuilder {
    constructor(options) {
        this._terminal = options.terminal;
        this._manifestsFilePath =
            options.manifestsFilePath || path.join(__dirname, 'manifestsFile', 'manifestsFile.js');
    }
    async buildManifestsJs(manifestsArray, outputFilePath) {
        // Write the manifests.js file
        const webpackConfig = this._getManifestsFileWebpackConfig(manifestsArray, outputFilePath);
        try {
            const stats = await node_core_library_1.LegacyAdapters.convertCallbackToPromise(webpack_1.default, [webpackConfig]);
            if (stats.hasErrors() || stats.hasWarnings()) {
                const statsJson = stats.toJson();
                const warnings = statsJson.warnings.join('\n');
                const errors = statsJson.errors.join('\n');
                const exceptionMessageParts = [];
                if (errors !== '') {
                    exceptionMessageParts.push(`errors: ${errors}`);
                }
                if (warnings !== '') {
                    exceptionMessageParts.push(`warnings: ${warnings}`);
                }
                throw new Error(exceptionMessageParts.join('\n'));
            }
        }
        catch (error) {
            this._terminal.writeErrorLine(`Error generating cumulative manifest: ${error}`);
            throw error;
        }
    }
    _getManifestsFileWebpackConfig(manifestsArray, outputFilePath) {
        const filename = path.basename(outputFilePath);
        return {
            mode: 'production',
            entry: this._manifestsFilePath,
            output: {
                library: 'debugManifests',
                libraryTarget: 'umd',
                path: path.dirname(outputFilePath),
                filename: filename
            },
            optimization: {
                minimize: false
            },
            resolve: {
                descriptionFiles: []
            },
            plugins: [
                new webpack_1.default.DefinePlugin({
                    MANIFESTS_ARRAY_PROXY: JSON.stringify(manifestsArray, undefined, 2),
                    MANIFESTS_FILE_NAME: JSON.stringify(filename)
                }),
                new set_webpack_public_path_plugin_1.SetPublicPathPlugin({
                    skipDetection: true,
                    scriptName: {
                        name: lodash_1.escapeRegExp(filename),
                        isTokenized: false
                    }
                })
            ],
            performance: {
                hints: false
            }
        };
    }
}
exports.ManifestsFileBuilder = ManifestsFileBuilder;
