// @microsoft/sp-loader/lib/DeveloperTools/Components/DeveloperModules/FlightInspector/FlightInspectorTab.js
import { __extends } from "tslib";
import * as React from 'react';
import styles from './FlightInspectorTab.module.scss';
import { SPComponentLoader } from '../../../../loader/SPComponentLoader';
var FlightInspectorTab = /** @class */ (function (_super) {
    __extends(FlightInspectorTab, _super);
    function FlightInspectorTab() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._rootRef = React.createRef();
        _this._handleCompare = function (theOtherFullString) {
            return _this._flightInspector.compare(theOtherFullString);
        };
        _this._loadFlightInspector = function () {
            void SPComponentLoader.loadComponentById('a7f8537f-f742-423f-ad20-ed8feaaf0805' /* sp-flight-inspector */).then(function (module) {
                var instance = module._SPFlightInspector.getInstance(_this.props.serviceScope);
                module._renderFlightViewer(_this._rootRef.current, {
                    enabledFlights: instance.getEnabledFlights(),
                    enabledExperiments: instance.getExperimentVariants(),
                    activatedKillSwitches: instance.getActivatedKillSwitches(),
                    fullString: instance.getFullString(),
                    onCompare: _this._handleCompare,
                    limitMaxSize: false
                });
                _this._flightInspector = instance;
            });
        };
        return _this;
    }
    FlightInspectorTab.prototype.componentDidMount = function () {
        this._loadFlightInspector();
    };
    FlightInspectorTab.prototype.render = function () {
        return (React.createElement("div", { className: styles.root },
            React.createElement("div", { className: styles.centerAlignedItem, ref: this._rootRef })));
    };
    return FlightInspectorTab;
}(React.Component));
export { FlightInspectorTab };
