// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/controls/TilesList/TilesList.module.scss.js
/* tslint:disable */
require("./TilesList.module.css");
var styles = {
    filePicker: 'filePicker_40fed829',
    tabsContainer: 'tabsContainer_40fed829',
    tabContainer: 'tabContainer_40fed829',
    nav: 'nav_40fed829',
    tabHeaderContainer: 'tabHeaderContainer_40fed829',
    breadcrumbNav: 'breadcrumbNav_40fed829',
    breadcrumbNavItem: 'breadcrumbNavItem_40fed829',
    tab: 'tab_40fed829',
    tabOffset: 'tabOffset_40fed829',
    tabFiles: 'tabFiles_40fed829',
    scrollablePaneWrapper: 'scrollablePaneWrapper_40fed829',
    actionButtonsContainer: 'actionButtonsContainer_40fed829',
    actionButtons: 'actionButtons_40fed829',
    actionButton: 'actionButton_40fed829',
    tabHeader: 'tabHeader_40fed829',
    selectedTab: 'selectedTab_40fed829',
    header: 'header_40fed829',
    xlg: 'xlg_40fed829',
    xxlg: 'xxlg_40fed829',
    xxxlg: 'xxxlg_40fed829',
    xxxxlg: 'xxxxlg_40fed829',
    itemPickerTopBar: 'itemPickerTopBar_40fed829',
    commandBarNoChevron: 'commandBarNoChevron_40fed829',
    fileItem: 'fileItem_40fed829',
    folderItem: 'folderItem_40fed829',
    fileTypeIcon: 'fileTypeIcon_40fed829',
    fileTypeIconIcon: 'fileTypeIconIcon_40fed829',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_40fed829',
    iconColumnHeader: 'iconColumnHeader_40fed829',
    fileRow: 'fileRow_40fed829',
    folderRow: 'folderRow_40fed829',
    emptyFolder: 'emptyFolder_40fed829',
    emptyFolderImage: 'emptyFolderImage_40fed829',
    emptyFolderImageTag: 'emptyFolderImageTag_40fed829',
    emptyFolderTitle: 'emptyFolderTitle_40fed829',
    emptyFolderSubText: 'emptyFolderSubText_40fed829',
    emptyFolderMobile: 'emptyFolderMobile_40fed829',
    emptyFolderPc: 'emptyFolderPc_40fed829',
    emptyFolderIcon: 'emptyFolderIcon_40fed829',
    aboveNameplate: 'aboveNameplate_40fed829',
    tileContent: 'tileContent_40fed829',
    foreground: 'foreground_40fed829',
    grid: 'grid_40fed829',
    cellContent: 'cellContent_40fed829',
    content: 'content_40fed829',
    tile: 'tile_40fed829',
    invokable: 'invokable_40fed829',
    namePlate: 'namePlate_40fed829',
    nameplate: 'nameplate_40fed829',
    name: 'name_40fed829',
    isLarge: 'isLarge_40fed829',
    activity: 'activity_40fed829',
    signalField: 'signalField_40fed829',
    compact: 'compact_40fed829',
    signalFieldValue: 'signalFieldValue_40fed829',
    folderList: 'folderList_40fed829',
    link: 'link_40fed829',
    cell: 'cell_40fed829',
    listCell: 'listCell_40fed829',
    label: 'label_40fed829',
    check: 'check_40fed829',
    selectable: 'selectable_40fed829',
    showCheck: 'showCheck_40fed829',
    selected: 'selected_40fed829',
    dialogMainOverride: 'dialogMainOverride_40fed829',
    itemTileBlankCover: 'itemTileBlankCover_40fed829',
    metadata: 'metadata_40fed829',
    isSmall: 'isSmall_40fed829',
    listPage: 'listPage_40fed829',
    odItemTile2FolderCover: 'odItemTile2FolderCover_40fed829',
    standaloneListBreadcrumb: 'standaloneListBreadcrumb_40fed829'
};
export default styles;
/* tslint:enable */
