// @microsoft/sp-loader/lib/debug/debugComponents/debugPrompt/DebugPrompt.js
import { __extends } from "tslib";
import * as React from 'react';
import { Dialog, DialogType, DialogFooter } from 'office-ui-fabric-react/lib/Dialog';
import { PrimaryButton, DefaultButton } from 'office-ui-fabric-react/lib/Button';
import { Callout, DirectionalHint } from 'office-ui-fabric-react/lib/Callout';
import { css } from 'office-ui-fabric-react/lib/Utilities';
import { Text } from '@microsoft/sp-core-library';
import * as strings from './DebugPrompt.resx';
import debugStyles from './../debugComponents.module.scss';
import styles from './DebugPrompt.module.scss';
var DebugPrompt = /** @class */ (function (_super) {
    __extends(DebugPrompt, _super);
    function DebugPrompt(props) {
        var _this = _super.call(this, props) || this;
        _this._handleCancelClick = function () {
            _this._handleButtonClick(false);
        };
        _this._handleConfirmClick = function () {
            _this._handleButtonClick(true);
        };
        _this._handleButtonClick = function (allow) {
            // Hide the dialog after 200ms to let the callout disappear
            _this._timeouts.push(setTimeout(function () { return _this.setState({ dismissed: true }); }, 200));
            _this.setState({
                showCallout: false,
                showDialog: false
            });
            _this.props.callback(allow);
        };
        _this.state = {
            showDialog: true,
            showCallout: false,
            dismissed: false
        };
        _this._calloutRef = React.createRef();
        _this._timeouts = [];
        return _this;
    }
    DebugPrompt.prototype.componentDidMount = function () {
        var _this = this;
        this._timeouts.push(setTimeout(function () {
            return _this.setState({
                showCallout: _this.state.showDialog // Only show the callout if the dialog is still open
            });
        }, 5000 // Show the callout after 5 seconds.
        ));
    };
    DebugPrompt.prototype.componentWillUnmount = function () {
        for (var _i = 0, _a = this._timeouts; _i < _a.length; _i++) {
            var id = _a[_i];
            clearTimeout(id);
        }
        this._timeouts = [];
    };
    DebugPrompt.prototype.render = function () {
        var _a;
        return !this.state.dismissed ? (React.createElement("div", { role: 'dialog', "data-automation-id": 'debugManifestPromptDialog', "aria-labelledby": 'manifestWarningLabel' },
            React.createElement(Dialog, { dialogContentProps: {
                    type: DialogType.largeHeader,
                    title: this.props.title
                }, hidden: !this.state.showDialog, modalProps: {
                    isBlocking: true,
                    containerClassName: debugStyles.dialogContainer
                } },
                React.createElement("p", { className: styles.warningText }, strings.debugManifestLoadingWarning),
                React.createElement("p", { className: styles.warningText }, Text.format(strings.debugManifestLoadingWarning2, strings.debugManifestLoadingCancel)),
                React.createElement(DialogFooter, null,
                    React.createElement(DefaultButton, { "data-automation-id": 'debugManifestConfirmButton', onClick: this._handleConfirmClick }, strings.debugManifestLoadingConfirm),
                    React.createElement(PrimaryButton, { "data-automation-id": 'debugManifestCancelButton', onClick: this._handleCancelClick },
                        React.createElement("div", { ref: this._calloutRef }, strings.debugManifestLoadingCancel)))),
            React.createElement(Callout, { className: css(styles.callout, (_a = {}, _a[styles.calloutFadeIn] = !!this.state.showCallout, _a)), target: this._calloutRef, directionalHint: DirectionalHint.bottomCenter, gapSpace: 10 },
                React.createElement("div", { className: styles.calloutText }, strings.debugManifestLoadingCalloutText)))) : null; // tslint:disable-line:no-null-keyword
    };
    return DebugPrompt;
}(React.Component));
export default DebugPrompt;
