// @microsoft/sp-component-base/lib-commonjs/index.js
"use strict";
/**
 * SharePoint Framework support for building client-side components
 *
 * @remarks
 * This package supports development of client-side components (e.g. web parts,
 * client-side extensions, and applications) built using the SharePoint Framework.
 * It defines the base classes and interfaces for functionality that is
 * common across all client-side component types.
 *
 * @packagedocumentation
 */
Object.defineProperty(exports, "__esModule", { value: true });
var BaseComponent_1 = require("./BaseComponent");
Object.defineProperty(exports, "BaseComponent", { enumerable: true, get: function () { return BaseComponent_1.default; } });
var BaseComponentContext_1 = require("./BaseComponentContext");
Object.defineProperty(exports, "BaseComponentContext", { enumerable: true, get: function () { return BaseComponentContext_1.default; } });
var DynamicDataProvider_1 = require("./DynamicDataProvider");
Object.defineProperty(exports, "DynamicDataProvider", { enumerable: true, get: function () { return DynamicDataProvider_1.default; } });
var DynamicDataSourceManager_1 = require("./DynamicDataSourceManager");
Object.defineProperty(exports, "DynamicDataSourceManager", { enumerable: true, get: function () { return DynamicDataSourceManager_1.default; } });
var DynamicProperty_1 = require("./DynamicProperty");
Object.defineProperty(exports, "DynamicProperty", { enumerable: true, get: function () { return DynamicProperty_1.default; } });
var LegacyThirdPartyFabricCoreLoader_1 = require("./LegacyThirdPartyFabricCoreLoader");
Object.defineProperty(exports, "_LegacyThirdPartyFabricCoreLoader", { enumerable: true, get: function () { return LegacyThirdPartyFabricCoreLoader_1.default; } });
var ThemeProvider_1 = require("./themeProvider/ThemeProvider");
Object.defineProperty(exports, "ThemeChangedEventArgs", { enumerable: true, get: function () { return ThemeProvider_1.ThemeChangedEventArgs; } });
Object.defineProperty(exports, "ThemeProvider", { enumerable: true, get: function () { return ThemeProvider_1.ThemeProvider; } });
// ShimmerFactory
var ShimmerFactory_1 = require("./ShimmerFactory");
Object.defineProperty(exports, "_ShimmerFactory", { enumerable: true, get: function () { return ShimmerFactory_1.default; } });
