// @microsoft/sp-http/lib/oauthTokenProvider/chunks/msalImplicit/MsalTokenProvider.js
import { __assign } from "tslib";
import { _QosMonitor } from '@microsoft/sp-diagnostics';
import AadTokenProvider from '../../AadTokenProvider';
import { BeforeRedirectEventArgs } from '../../AadTokenProvider';
import { Guid, _BrowserDetection, _Browser, _SPEventManager, _SPFlight } from '@microsoft/sp-core-library';
import AadConstants from '../../AadConstants';
import { AadErrorHandler } from '../../AadErrorHandler';
import { AadKillSwitches } from '../../AadKillSwitches';
import * as msal from 'msal';
/**
 * Class that wraps the MSAL's authentication class
 * @internal
 */
var MsalTokenProvider = /** @class */ (function () {
    function MsalTokenProvider(configuration) {
        this._defaultConfiguration = configuration;
        this._requestMap = new Map();
        this._msalInstance = this._initializeMsalJs();
        // Revert to old behavior if Flight is off
        if (!_SPFlight.isEnabled(1887) /* Use MSAL for 3rd Party Requests */) {
            this._defaultConfiguration.aadInstanceUrl = 'https://login.microsoftonline.com';
        }
    }
    MsalTokenProvider.prototype.getToken = function (resourceEndpoint, useCachedToken) {
        var _this = this;
        if (useCachedToken === void 0) { useCachedToken = true; }
        var acquireAccessTokenQosMonitor = new _QosMonitor('MsalTokenProvider.GetToken');
        var extraData = this._generateTelemetryExtraData();
        if (AadKillSwitches.isLogAadAdditionalTelemetryEnabled() && extraData.isInternal) {
            extraData.name = resourceEndpoint;
        }
        // Supporting legacy behavor when the MSAL flight is off
        if (!_SPFlight.isEnabled(1887) /* Use MSAL for 3rd Party Requests */) {
            // MSAL will return "user login is required" in this case.
            // This will reduce the noise in the MSAL.AcquireAccessTokenSilent monitor.
            if (!this._defaultConfiguration.userPrincipalName) {
                var missingLoginHintError = new Error('MSAL does not support silent auth without login hint');
                acquireAccessTokenQosMonitor.writeExpectedFailure(missingLoginHintError.message, missingLoginHintError, extraData);
                return Promise.reject(missingLoginHintError);
            }
        }
        // If you try to fetch a token without "logging in", MSAL will make a request to fetch both
        // an ID and ACCESS token in parallel. The problem with this is that the URL containing the tokens
        // will exceed 4K characters. IE and EDGE do not support this and will automatically drop off
        // trailing query parameters. MSAL team plans to eventually fix this on their end.
        var currentBrowser = _BrowserDetection.getBrowserInformation().browser;
        var acquireTokenPromise;
        if (currentBrowser === _Browser.IE || currentBrowser === _Browser.Edge) {
            acquireTokenPromise = this._loginSilent().then(function () { return _this._acquireToken(resourceEndpoint); });
        }
        else {
            acquireTokenPromise = this._acquireToken(resourceEndpoint);
        }
        extraData.CorrelationId = this._defaultConfiguration.spRequestGuid || Guid.newGuid().toString();
        return acquireTokenPromise
            .then(function (accessToken) {
            acquireAccessTokenQosMonitor.writeSuccess(extraData);
            return accessToken;
        })
            .catch(function (e) { return _this._handleAuthErrors(acquireAccessTokenQosMonitor, e, extraData, resourceEndpoint); });
    };
    MsalTokenProvider.prototype._loginSilent = function () {
        // Represents the time it takes to fetch an id token.
        var getIdTokenQosMonitor = new _QosMonitor('MsalTokenProvider.GetIdToken');
        return this._loginSilentHelper(0, getIdTokenQosMonitor);
    };
    MsalTokenProvider.prototype._loginSilentHelper = function (attempt, getIdTokenQosMonitor) {
        var _this = this;
        var retryQosMonitor;
        if (attempt > 0) {
            retryQosMonitor = new _QosMonitor('MsalTokenProvider.RetryGetIdToken');
        }
        var extraData = this._generateTelemetryExtraData();
        var requestData = {
            authority: this._defaultConfiguration.aadInstanceUrl + '/' + this._defaultConfiguration.aadTenantId,
            correlationId: extraData.CorrelationId,
            loginHint: this._defaultConfiguration.userPrincipalName,
            scopes: [this._defaultConfiguration.servicePrincipalId]
        };
        if (_SPFlight.isEnabled(1887) /* Use MSAL for 3rd Party Requests */ && attempt === 0) {
            requestData.sid = this._defaultConfiguration.aadSessionId;
            requestData.loginHint = undefined;
        }
        return this._msalInstance
            .acquireTokenSilent(requestData)
            .then(function (response) {
            // Using alias column to mark cache hits
            extraData.alias = response.fromCache.toString();
            if (AadKillSwitches.isAdditionalMsalTelemetryEnabled()) {
                extraData.jsonExtraData = JSON.stringify(_this._requestMap.get(extraData.CorrelationId));
            }
            getIdTokenQosMonitor.writeSuccess(extraData);
            return response.accessToken;
        })
            .catch(function (e) {
            if (AadKillSwitches.isAdditionalMsalTelemetryEnabled()) {
                extraData.jsonExtraData = JSON.stringify(_this._requestMap.get(extraData.CorrelationId));
            }
            retryQosMonitor === null || retryQosMonitor === void 0 ? void 0 : retryQosMonitor.writeUnexpectedFailure(e.errorCode, e, extraData);
            if (AadKillSwitches.isRetryIdTokenWithLoginHintEnabled() &&
                AadErrorHandler._isInteractionRequired(e.message, e.errorCode) &&
                attempt < 1) {
                return _this._loginSilentHelper(++attempt, getIdTokenQosMonitor);
            }
            else {
                _this._handleAuthErrors(getIdTokenQosMonitor, e, extraData, _this._defaultConfiguration.servicePrincipalId);
            }
        });
    };
    MsalTokenProvider.prototype._acquireToken = function (resourceEndpoint) {
        // Represents the time it takes to fetch an access token.
        var getAccessTokenQosMonitor = new _QosMonitor('MsalTokenProvider.GetAccessToken');
        return this._acquireTokenHelper(resourceEndpoint, 0, getAccessTokenQosMonitor);
    };
    MsalTokenProvider.prototype._acquireTokenHelper = function (resourceEndpoint, attempt, getAccessTokenQosMonitor, useLoginHint) {
        var _this = this;
        if (useLoginHint === void 0) { useLoginHint = false; }
        var retryQosMonitor;
        if (attempt > 0) {
            if (useLoginHint) {
                retryQosMonitor = new _QosMonitor('MsalTokenProvider.RetryLoginHintGetAccessToken');
            }
            else {
                retryQosMonitor = new _QosMonitor('MsalTokenProvider.RetryGetAccessToken');
            }
        }
        var extraData = this._generateTelemetryExtraData();
        var requestData = {
            authority: this._defaultConfiguration.aadInstanceUrl + '/' + this._defaultConfiguration.aadTenantId,
            correlationId: extraData.CorrelationId,
            loginHint: this._defaultConfiguration.userPrincipalName,
            scopes: [resourceEndpoint + '/.default']
        };
        if (_SPFlight.isEnabled(1887) /* Use MSAL for 3rd Party Requests */ && !useLoginHint) {
            requestData.sid = this._defaultConfiguration.aadSessionId;
            requestData.loginHint = undefined;
        }
        return this._msalInstance
            .acquireTokenSilent(requestData)
            .then(function (response) {
            // Using alias column to mark cache hits
            extraData.alias = response.fromCache.toString();
            if (AadKillSwitches.isAdditionalMsalTelemetryEnabled()) {
                extraData.jsonExtraData = JSON.stringify(_this._requestMap.get(extraData.CorrelationId));
            }
            getAccessTokenQosMonitor.writeSuccess(extraData);
            retryQosMonitor === null || retryQosMonitor === void 0 ? void 0 : retryQosMonitor.writeSuccess(extraData);
            return response.accessToken;
        })
            .catch(function (e) {
            if (AadKillSwitches.isAdditionalMsalTelemetryEnabled()) {
                extraData.jsonExtraData = JSON.stringify(_this._requestMap.get(extraData.CorrelationId));
            }
            retryQosMonitor === null || retryQosMonitor === void 0 ? void 0 : retryQosMonitor.writeUnexpectedFailure(e.errorCode, e, extraData);
            // We only want to retry once for network issues as the failure could be due to many different reasons
            // and retry could help fix the issue
            if ((AadKillSwitches.isRetryTimeoutsEnabled() &&
                AadErrorHandler._isTokenRenewalTimeout(e.errorCode) &&
                attempt < 1) ||
                (AadKillSwitches.isRetryOnResolveEndpointsFailureEnabled() &&
                    AadErrorHandler._isEndpointsResolutionError(e.errorCode) &&
                    attempt < 1)) {
                return _this._acquireTokenHelper(resourceEndpoint, ++attempt, getAccessTokenQosMonitor);
            }
            else if (AadKillSwitches.isRetryWithLoginHintEnabled() &&
                AadErrorHandler._isInteractionRequired(e.message, e.errorCode) &&
                attempt < 1) {
                return _this._acquireTokenHelper(resourceEndpoint, ++attempt, getAccessTokenQosMonitor, true);
            }
            else {
                _this._handleAuthErrors(getAccessTokenQosMonitor, e, extraData, resourceEndpoint);
            }
        });
    };
    MsalTokenProvider.prototype._initializeMsalJs = function () {
        var _this = this;
        var configuration;
        if (_SPFlight.isEnabled(467)) {
            configuration = {
                auth: {
                    clientId: this._defaultConfiguration.servicePrincipalId,
                    navigateToLoginRequestUrl: true,
                    redirectUri: this._defaultConfiguration.redirectUri,
                    validateAuthority: false,
                    authority: this._defaultConfiguration.aadInstanceUrl + '/' + this._defaultConfiguration.aadTenantId
                },
                system: {
                    loadFrameTimeout: 10000,
                    navigateFrameWait: 0
                }
            };
        }
        else {
            configuration = {
                auth: {
                    clientId: this._defaultConfiguration.servicePrincipalId,
                    navigateToLoginRequestUrl: false,
                    redirectUri: this._defaultConfiguration.redirectUri,
                    validateAuthority: false
                },
                system: {
                    loadFrameTimeout: 10000,
                    navigateFrameWait: 0
                }
            };
        }
        if (AadKillSwitches.isAdditionalMsalTelemetryEnabled()) {
            configuration.system.telemetry = {
                applicationName: this._defaultConfiguration.servicePrincipalId,
                applicationVersion: '1.0',
                telemetryEmitter: function (events) {
                    var key = '';
                    var value = {};
                    events.forEach(function (event) {
                        if (event['msal.event_name'] === 'msal.api_event') {
                            key = event['Microsoft.MSAL.correlation_id'];
                            value = __assign(__assign(__assign({}, value), _this._requestMap.get(key)), event);
                        }
                        else if (event['msal.event_name'] === 'msal.default_event') {
                            value = __assign(__assign({}, value), { 'msal.effective_connection_speed': event['msal.effective_connection_speed'] });
                        }
                    });
                    _this._requestMap.set(key, value);
                }
            };
        }
        return new msal.UserAgentApplication(configuration);
    };
    MsalTokenProvider.prototype._generateTelemetryExtraData = function () {
        // alias is being used to detect cache hits
        // manifestId is used to save msal perf data
        return {
            alias: 'false',
            CorrelationId: Guid.newGuid().toString(),
            isInternal: this._defaultConfiguration.servicePrincipalId === AadConstants.PRE_AUTHORIZED_APP_PRINCIPAL_ID,
            jsonExtraData: ''
        };
    };
    MsalTokenProvider.prototype._cancelRedirect = function () {
        this._redirectCancelled = true;
    };
    MsalTokenProvider.prototype._handleAuthErrors = function (monitor, error, extraData, resourceEndpoint) {
        var _this = this;
        var userAgentString = _BrowserDetection.getBrowserInformation().userAgent || '';
        // Will remove this flight check when MSAL provides an API for handling MFA.
        if (_SPFlight.isEnabled(467) /* Allow MSAL to prompt the end user for MFA errors */ &&
            AadErrorHandler._isInteractionRequired(error.message, error.errorCode) &&
            _Browser.IE !== _BrowserDetection.getBrowserInformation().browser) {
            // To prevent infinite redirect loops we want to only allow 3 redirects with 1 minute
            var currDate = new Date();
            var firstRedirectTime = window.sessionStorage.getItem('msalRedirectFirstAttempTime');
            var oneMinuteInMs = 60 * 1000;
            var foundPreviousAttempt = false;
            var attemptNumber = 1;
            if (firstRedirectTime) {
                var firstRedirectDate = new Date(firstRedirectTime);
                if (currDate.getTime() - firstRedirectDate.getTime() < oneMinuteInMs) {
                    var attemptCount = window.sessionStorage.getItem('msalRedirectAttempNumber');
                    if (attemptCount) {
                        attemptNumber = parseInt(attemptCount, 10);
                        attemptNumber++;
                        foundPreviousAttempt = true;
                    }
                }
            }
            if (!foundPreviousAttempt) {
                window.sessionStorage.setItem('msalRedirectFirstAttempTime', currDate.toString());
            }
            // We want a maxium of 3 attempts, and have already incremented from 1 for previous attempts
            if (attemptNumber < 4) {
                monitor.writeExpectedFailure(error.errorCode, error, extraData);
                window.sessionStorage.setItem('msalRedirectAttempNumber', attemptNumber.toString());
                window.sessionStorage.setItem('msalRedirectClientId', this._defaultConfiguration.servicePrincipalId);
                this._msalInstance.acquireTokenRedirect({
                    authority: this._defaultConfiguration.aadInstanceUrl + '/' + this._defaultConfiguration.aadTenantId,
                    correlationId: extraData.CorrelationId,
                    loginHint: this._defaultConfiguration.userPrincipalName,
                    scopes: [resourceEndpoint + '/.default'],
                    state: window.location.href,
                    onRedirectNavigate: function (url) {
                        _this._redirectCancelled = false;
                        _SPEventManager.instance.raiseEvent(AadTokenProvider._onBeforeRedirectEventId, new BeforeRedirectEventArgs(url, _this._cancelRedirect.bind(_this)));
                        // If we are in tab cancel the redirect.
                        if (window.sessionStorage.getItem('_isRunningTABTest') === 'true') {
                            _this._redirectCancelled = true;
                        }
                        return !_this._redirectCancelled;
                    }
                });
            }
            else {
                monitor.writeUnexpectedFailure(error.errorCode, error, extraData);
            }
        }
        else if (AadErrorHandler._doesAadErrorCodeExist(error.message, AadConstants.EXPECTED_AAD_ERRORS) ||
            userAgentString.indexOf('TabStop/1.0') > -1 // MSAL Implicit seems to have issues with test infra.
        ) {
            monitor.writeExpectedFailure(error.errorCode, error, extraData);
        }
        else {
            monitor.writeUnexpectedFailure(error.errorCode, error, extraData);
        }
        throw error;
    };
    return MsalTokenProvider;
}());
export { MsalTokenProvider };
