// @microsoft/rush-lib/lib/logic/deploy/DeployArchiver.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeployArchiver = void 0;
const JSZip = require("jszip");
const path = __importStar(require("path"));
const node_core_library_1 = require("@rushstack/node-core-library");
class DeployArchiver {
    static async createArchiveAsync(deployState) {
        if (deployState.createArchiveFilePath !== undefined) {
            console.log('Creating archive...');
            const zip = this._getZipOfFolder(deployState.targetRootFolder);
            const zipContent = await zip.generateAsync({
                type: 'nodebuffer',
                platform: 'UNIX'
            });
            node_core_library_1.FileSystem.writeFile(path.resolve(deployState.targetRootFolder, deployState.createArchiveFilePath), zipContent);
            console.log('Archive created successfully.');
        }
    }
    static _getFilePathsRecursively(dir) {
        // returns a flat array of absolute paths of all files recursively contained in the dir
        let results = [];
        const list = node_core_library_1.FileSystem.readFolder(dir);
        if (!list.length)
            return results;
        for (let file of list) {
            file = path.resolve(dir, file);
            const stat = node_core_library_1.FileSystem.getLinkStatistics(file);
            if (stat && stat.isDirectory()) {
                results = results.concat(this._getFilePathsRecursively(file));
            }
            else {
                results.push(file);
            }
        }
        return results;
    }
    static _getZipOfFolder(dir) {
        // returns a JSZip instance filled with contents of dir.
        const allPaths = this._getFilePathsRecursively(dir);
        // This value sets the allowed permissions when preserving symbolic links.
        // 120000 is the symbolic link identifier, and 0755 designates the allowed permissions.
        // See: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/stat.h#n10
        const permissionsValue = 0o120755;
        const zip = new JSZip();
        for (const filePath of allPaths) {
            // Get the relative path and replace backslashes for Unix compat
            const addPath = node_core_library_1.Path.convertToSlashes(path.relative(dir, filePath));
            const stat = node_core_library_1.FileSystem.getLinkStatistics(filePath);
            const permissions = stat.mode;
            if (stat.isSymbolicLink()) {
                zip.file(addPath, node_core_library_1.FileSystem.readLink(filePath), {
                    unixPermissions: permissionsValue,
                    dir: stat.isDirectory()
                });
            }
            else {
                const data = node_core_library_1.FileSystem.readFileToBuffer(filePath);
                zip.file(addPath, data, {
                    unixPermissions: permissions,
                    dir: stat.isDirectory()
                });
            }
        }
        return zip;
    }
}
exports.DeployArchiver = DeployArchiver;
