// @microsoft/rush-lib/lib/logic/deploy/SymlinkAnalyzer.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SymlinkAnalyzer = void 0;
const node_core_library_1 = require("@rushstack/node-core-library");
const path = __importStar(require("path"));
class SymlinkAnalyzer {
    constructor() {
        // The directory tree discovered so far
        this._nodesByPath = new Map();
        // The symlinks that we encountered while building the directory tree
        this._linkInfosByPath = new Map();
    }
    analyzePath(inputPath, preserveLinks = false) {
        let pathSegments = path.resolve(inputPath).split(path.sep);
        let pathSegmentsIndex = 0;
        for (;;) {
            const currentPath = pathSegments.slice(0, pathSegmentsIndex + 1).join(path.sep);
            if (currentPath === '') {
                // Edge case for a Unix path like "/folder/file" --> [ "", "folder", "file" ]
                ++pathSegmentsIndex;
                continue;
            }
            let currentNode = this._nodesByPath.get(currentPath);
            if (currentNode === undefined) {
                const linkStats = node_core_library_1.FileSystem.getLinkStatistics(currentPath);
                if (linkStats.isSymbolicLink()) {
                    const linkTargetPath = node_core_library_1.FileSystem.readLink(currentPath);
                    const parentFolder = path.join(currentPath, '..');
                    const resolvedLinkTargetPath = path.resolve(parentFolder, linkTargetPath);
                    currentNode = {
                        kind: 'link',
                        nodePath: currentPath,
                        linkTarget: resolvedLinkTargetPath
                    };
                }
                else if (linkStats.isDirectory()) {
                    currentNode = {
                        kind: 'folder',
                        nodePath: currentPath
                    };
                }
                else if (linkStats.isFile()) {
                    currentNode = {
                        kind: 'file',
                        nodePath: currentPath
                    };
                }
                else {
                    throw new Error('Unknown object type: ' + currentPath);
                }
                this._nodesByPath.set(currentPath, currentNode);
            }
            ++pathSegmentsIndex;
            if (!preserveLinks) {
                while (currentNode.kind === 'link') {
                    const targetNode = this.analyzePath(currentNode.linkTarget, true);
                    // Have we created an ILinkInfo for this link yet?
                    if (!this._linkInfosByPath.has(currentNode.nodePath)) {
                        // Follow any symbolic links to determine whether the final target is a directory
                        const targetIsDirectory = node_core_library_1.FileSystem.getStatistics(targetNode.nodePath).isDirectory();
                        const linkInfo = {
                            kind: targetIsDirectory ? 'folderLink' : 'fileLink',
                            linkPath: currentNode.nodePath,
                            targetPath: targetNode.nodePath
                        };
                        this._linkInfosByPath.set(currentNode.nodePath, linkInfo);
                    }
                    const targetSegments = targetNode.nodePath.split(path.sep);
                    const remainingSegments = pathSegments.slice(pathSegmentsIndex);
                    pathSegments = [...targetSegments, ...remainingSegments];
                    pathSegmentsIndex = targetSegments.length;
                    currentNode = targetNode;
                }
            }
            if (pathSegmentsIndex >= pathSegments.length) {
                // We reached the end
                return currentNode;
            }
            if (currentNode.kind !== 'folder') {
                // This should never happen, because analyzePath() is always supposed to receive complete paths
                // to real filesystem objects.
                throw new node_core_library_1.InternalError('The path ends prematurely at: ' + inputPath);
            }
        }
    }
    /**
     * Returns a summary of all the symbolic links encountered by {@link SymlinkAnalyzer.analyzePath}.
     */
    reportSymlinks() {
        const list = [...this._linkInfosByPath.values()];
        node_core_library_1.Sort.sortBy(list, (x) => x.linkPath);
        return list;
    }
}
exports.SymlinkAnalyzer = SymlinkAnalyzer;
